// @serverless-devs/core/dist/common/execCommand/component/hook.js
"use strict";
var __assign = (this && this.__assign) || function () {
    __assign = Object.assign || function(t) {
        for (var s, i = 1, n = arguments.length; i < n; i++) {
            s = arguments[i];
            for (var p in s) if (Object.prototype.hasOwnProperty.call(s, p))
                t[p] = s[p];
        }
        return t;
    };
    return __assign.apply(this, arguments);
};
var __awaiter = (this && this.__awaiter) || function (thisArg, _arguments, P, generator) {
    function adopt(value) { return value instanceof P ? value : new P(function (resolve) { resolve(value); }); }
    return new (P || (P = Promise))(function (resolve, reject) {
        function fulfilled(value) { try { step(generator.next(value)); } catch (e) { reject(e); } }
        function rejected(value) { try { step(generator["throw"](value)); } catch (e) { reject(e); } }
        function step(result) { result.done ? resolve(result.value) : adopt(result.value).then(fulfilled, rejected); }
        step((generator = generator.apply(thisArg, _arguments || [])).next());
    });
};
var __generator = (this && this.__generator) || function (thisArg, body) {
    var _ = { label: 0, sent: function() { if (t[0] & 1) throw t[1]; return t[1]; }, trys: [], ops: [] }, f, y, t, g;
    return g = { next: verb(0), "throw": verb(1), "return": verb(2) }, typeof Symbol === "function" && (g[Symbol.iterator] = function() { return this; }), g;
    function verb(n) { return function (v) { return step([n, v]); }; }
    function step(op) {
        if (f) throw new TypeError("Generator is already executing.");
        while (_) try {
            if (f = 1, y && (t = op[0] & 2 ? y["return"] : op[0] ? y["throw"] || ((t = y["return"]) && t.call(y), 0) : y.next) && !(t = t.call(y, op[1])).done) return t;
            if (y = 0, t) op = [op[0] & 2, t.value];
            switch (op[0]) {
                case 0: case 1: t = op; break;
                case 4: _.label++; return { value: op[1], done: false };
                case 5: _.label++; y = op[1]; op = [0]; continue;
                case 7: op = _.ops.pop(); _.trys.pop(); continue;
                default:
                    if (!(t = _.trys, t = t.length > 0 && t[t.length - 1]) && (op[0] === 6 || op[0] === 2)) { _ = 0; continue; }
                    if (op[0] === 3 && (!t || (op[1] > t[0] && op[1] < t[3]))) { _.label = op[1]; break; }
                    if (op[0] === 6 && _.label < t[1]) { _.label = t[1]; t = op; break; }
                    if (t && _.label < t[2]) { _.label = t[2]; _.ops.push(op); break; }
                    if (t[2]) _.ops.pop();
                    _.trys.pop(); continue;
            }
            op = body.call(thisArg, _);
        } catch (e) { op = [6, e]; y = 0; } finally { f = t = 0; }
        if (op[0] & 5) throw op[1]; return { value: op[0] ? op[1] : void 0, done: true };
    }
};
var __importDefault = (this && this.__importDefault) || function (mod) {
    return (mod && mod.__esModule) ? mod : { "default": mod };
};
Object.defineProperty(exports, "__esModule", { value: true });
var fs_extra_1 = __importDefault(require("fs-extra"));
var logger_1 = require("../../../logger");
var execa_1 = __importDefault(require("execa"));
var lodash_1 = require("lodash");
var libs_1 = require("../../../libs");
var load_1 = require("../../load");
var utils_1 = require("../utils");
var chalk_1 = __importDefault(require("chalk"));
var os_1 = __importDefault(require("os"));
var string_argv_1 = __importDefault(require("string-argv"));
var error_1 = require("../../error");
var Hook = /** @class */ (function () {
    function Hook(inputs) {
        this.inputs = inputs;
        this.preHooks = [];
        this.afterHooks = [];
    }
    Hook.prototype.init = function (hooks) {
        if (hooks === void 0) { hooks = []; }
        this.preHooks = [];
        this.afterHooks = [];
        for (var _i = 0, hooks_1 = hooks; _i < hooks_1.length; _i++) {
            var hook = hooks_1[_i];
            hook.pre ? this.preHooks.push(hook) : this.afterHooks.push(hook);
        }
        return this;
    };
    Hook.prototype.executePreHook = function () {
        return __awaiter(this, void 0, void 0, function () {
            var temp, _i, _a, hook;
            return __generator(this, function (_b) {
                switch (_b.label) {
                    case 0:
                        if (!(this.preHooks.length > 0)) return [3 /*break*/, 5];
                        logger_1.logger.info('Start the pre-action');
                        _i = 0, _a = this.preHooks;
                        _b.label = 1;
                    case 1:
                        if (!(_i < _a.length)) return [3 /*break*/, 4];
                        hook = _a[_i];
                        logger_1.logger.info("Action: " + hook.value);
                        return [4 /*yield*/, this.commandExecute(hook)];
                    case 2:
                        temp = _b.sent();
                        _b.label = 3;
                    case 3:
                        _i++;
                        return [3 /*break*/, 1];
                    case 4:
                        logger_1.logger.info('End the pre-action');
                        _b.label = 5;
                    case 5: return [2 /*return*/, temp];
                }
            });
        });
    };
    Hook.prototype.executeAfterHook = function (_a) {
        var output = _a.output;
        return __awaiter(this, void 0, void 0, function () {
            var _i, _b, hook;
            return __generator(this, function (_c) {
                switch (_c.label) {
                    case 0:
                        if (!(this.afterHooks.length > 0)) return [3 /*break*/, 5];
                        this.output = output;
                        logger_1.logger.info('Start the after-action');
                        _i = 0, _b = this.afterHooks;
                        _c.label = 1;
                    case 1:
                        if (!(_i < _b.length)) return [3 /*break*/, 4];
                        hook = _b[_i];
                        logger_1.logger.info("Action: " + hook.value);
                        return [4 /*yield*/, this.commandExecute(hook)];
                    case 2:
                        _c.sent();
                        _c.label = 3;
                    case 3:
                        _i++;
                        return [3 /*break*/, 1];
                    case 4:
                        logger_1.logger.info('End the after-action');
                        _c.label = 5;
                    case 5: return [2 /*return*/];
                }
            });
        });
    };
    Hook.prototype.commandExecute = function (configs) {
        return __awaiter(this, void 0, void 0, function () {
            var execPath, argv, result, instance, result;
            return __generator(this, function (_a) {
                switch (_a.label) {
                    case 0:
                        if (configs.type === 'run') {
                            execPath = configs.path;
                            if (fs_extra_1.default.existsSync(execPath) && fs_extra_1.default.lstatSync(execPath).isDirectory()) {
                                try {
                                    execa_1.default.sync(configs.value, {
                                        cwd: execPath,
                                        stdio: 'inherit',
                                        shell: true,
                                    });
                                }
                                catch (error) {
                                    if (os_1.default.platform() === 'win32') {
                                        logger_1.logger.info('Command run execution environment：CMD');
                                        new error_1.HumanWarning({
                                            warningMessage: 'Please check whether the actions section of yaml can be executed in the current environment.',
                                        });
                                    }
                                    utils_1.throwError({
                                        error: error,
                                        serviceName: lodash_1.get(this.inputs, 'project.projectName'),
                                    });
                                }
                            }
                        }
                        if (!(configs.type === 'component')) return [3 /*break*/, 2];
                        argv = string_argv_1.default(configs.value);
                        return [4 /*yield*/, this.execComponent({ argv: argv })];
                    case 1:
                        result = _a.sent();
                        return [2 /*return*/, {
                                type: configs.type,
                                data: result,
                            }];
                    case 2:
                        if (!(configs.type === 'plugin')) return [3 /*break*/, 5];
                        return [4 /*yield*/, load_1.loadComponent(configs.value)];
                    case 3:
                        instance = _a.sent();
                        return [4 /*yield*/, instance(__assign(__assign({}, this.inputs), { output: this.output }), configs.args)];
                    case 4:
                        result = _a.sent();
                        return [2 /*return*/, {
                                type: configs.type,
                                data: result,
                            }];
                    case 5: return [2 /*return*/];
                }
            });
        });
    };
    Hook.prototype.execComponent = function (_a) {
        var argv = _a.argv;
        return __awaiter(this, void 0, void 0, function () {
            var rawData, componentName, method, argsObj, instance, newInputs, _b, error_2;
            return __generator(this, function (_c) {
                switch (_c.label) {
                    case 0:
                        rawData = libs_1.getGlobalArgs(argv)._;
                        componentName = rawData[0], method = rawData[1];
                        argsObj = lodash_1.filter(argv.slice(2), function (o) { return !lodash_1.includes([componentName, method], o); });
                        if (lodash_1.isEmpty(method))
                            return [2 /*return*/];
                        return [4 /*yield*/, load_1.loadComponent(componentName)];
                    case 1:
                        instance = _c.sent();
                        if (!instance[method]) return [3 /*break*/, 5];
                        _c.label = 2;
                    case 2:
                        _c.trys.push([2, 4, , 5]);
                        newInputs = __assign(__assign({}, this.inputs), { args: lodash_1.join(argsObj, ' '), argsObj: argsObj, output: this.output });
                        _b = this;
                        return [4 /*yield*/, instance[method](newInputs)];
                    case 3:
                        _b.output = _c.sent();
                        return [2 /*return*/, this.output];
                    case 4:
                        error_2 = _c.sent();
                        utils_1.throwError({
                            error: error_2,
                            serviceName: lodash_1.get(this.inputs, 'project.projectName'),
                        });
                        return [3 /*break*/, 5];
                    case 5: 
                    // 方法不存在，此时系统将会认为是未找到组件方法，系统的exit code为100；
                    throw new Error(JSON.stringify({
                        code: 100,
                        message: "The [" + method + "] command was not found.",
                        tips: "Please check the component " + componentName + " has the " + method + " method. Serverless Devs documents\uFF1A" + chalk_1.default.underline('https://github.com/Serverless-Devs/Serverless-Devs/blob/master/docs/zh/command'),
                    }));
                }
            });
        });
    };
    return Hook;
}());
exports.default = Hook;
