// @serverless-devs/core/dist/common/execCommand/component/index.js
"use strict";
var __awaiter = (this && this.__awaiter) || function (thisArg, _arguments, P, generator) {
    function adopt(value) { return value instanceof P ? value : new P(function (resolve) { resolve(value); }); }
    return new (P || (P = Promise))(function (resolve, reject) {
        function fulfilled(value) { try { step(generator.next(value)); } catch (e) { reject(e); } }
        function rejected(value) { try { step(generator["throw"](value)); } catch (e) { reject(e); } }
        function step(result) { result.done ? resolve(result.value) : adopt(result.value).then(fulfilled, rejected); }
        step((generator = generator.apply(thisArg, _arguments || [])).next());
    });
};
var __generator = (this && this.__generator) || function (thisArg, body) {
    var _ = { label: 0, sent: function() { if (t[0] & 1) throw t[1]; return t[1]; }, trys: [], ops: [] }, f, y, t, g;
    return g = { next: verb(0), "throw": verb(1), "return": verb(2) }, typeof Symbol === "function" && (g[Symbol.iterator] = function() { return this; }), g;
    function verb(n) { return function (v) { return step([n, v]); }; }
    function step(op) {
        if (f) throw new TypeError("Generator is already executing.");
        while (_) try {
            if (f = 1, y && (t = op[0] & 2 ? y["return"] : op[0] ? y["throw"] || ((t = y["return"]) && t.call(y), 0) : y.next) && !(t = t.call(y, op[1])).done) return t;
            if (y = 0, t) op = [op[0] & 2, t.value];
            switch (op[0]) {
                case 0: case 1: t = op; break;
                case 4: _.label++; return { value: op[1], done: false };
                case 5: _.label++; y = op[1]; op = [0]; continue;
                case 7: op = _.ops.pop(); _.trys.pop(); continue;
                default:
                    if (!(t = _.trys, t = t.length > 0 && t[t.length - 1]) && (op[0] === 6 || op[0] === 2)) { _ = 0; continue; }
                    if (op[0] === 3 && (!t || (op[1] > t[0] && op[1] < t[3]))) { _.label = op[1]; break; }
                    if (op[0] === 6 && _.label < t[1]) { _.label = t[1]; t = op; break; }
                    if (t && _.label < t[2]) { _.label = t[2]; _.ops.push(op); break; }
                    if (t[2]) _.ops.pop();
                    _.trys.pop(); continue;
            }
            op = body.call(thisArg, _);
        } catch (e) { op = [6, e]; y = 0; } finally { f = t = 0; }
        if (op[0] & 5) throw op[1]; return { value: op[0] ? op[1] : void 0, done: true };
    }
};
var __importDefault = (this && this.__importDefault) || function (mod) {
    return (mod && mod.__esModule) ? mod : { "default": mod };
};
Object.defineProperty(exports, "__esModule", { value: true });
var libs_1 = require("../../../libs");
var path_1 = __importDefault(require("path"));
var credential_1 = require("../../credential");
var utils_1 = require("../utils");
var hook_1 = __importDefault(require("./hook"));
var load_1 = require("../../load");
var constant_1 = require("../../constant");
var chalk_1 = __importDefault(require("chalk"));
var lodash_1 = require("lodash");
var logger_1 = require("../../../logger");
var report_1 = __importDefault(require("../../report"));
var lodash_2 = require("lodash");
var constant_2 = require("../../constant");
var ComponentExec = /** @class */ (function () {
    function ComponentExec(config) {
        this.config = config;
        this.projectConfig = config.projectConfig;
    }
    ComponentExec.prototype.handleCredentials = function () {
        return __awaiter(this, void 0, void 0, function () {
            var projectConfig, _a, _b, _c, accessPath, data, credentials, accessFromEnv;
            var _d;
            return __generator(this, function (_e) {
                switch (_e.label) {
                    case 0:
                        projectConfig = this.config.projectConfig;
                        if (!(projectConfig.access === constant_2.ALIYUN_CLI)) return [3 /*break*/, 2];
                        _a = this;
                        _b = lodash_1.assign;
                        _c = [{}, projectConfig];
                        _d = {};
                        return [4 /*yield*/, credential_1.getCredential(projectConfig.access)];
                    case 1: return [2 /*return*/, (_a.projectConfig = _b.apply(void 0, _c.concat([(_d.credentials = _e.sent(),
                                _d)])))];
                    case 2:
                        accessPath = path_1.default.join(libs_1.getRootHome(), 'access.yaml');
                        return [4 /*yield*/, libs_1.getYamlContent(accessPath)];
                    case 3:
                        data = _e.sent();
                        if (!lodash_2.get(data, projectConfig.access)) return [3 /*break*/, 5];
                        return [4 /*yield*/, credential_1.getCredential(projectConfig.access)];
                    case 4:
                        credentials = _e.sent();
                        this.projectConfig = lodash_1.assign({}, projectConfig, { credentials: credentials });
                        _e.label = 5;
                    case 5: return [4 /*yield*/, credential_1.getCredentialFromEnv(projectConfig.access)];
                    case 6:
                        accessFromEnv = _e.sent();
                        if (accessFromEnv) {
                            this.projectConfig = lodash_1.assign({}, projectConfig, { credentials: accessFromEnv });
                        }
                        return [2 /*return*/];
                }
            });
        });
    };
    ComponentExec.prototype.init = function () {
        return __awaiter(this, void 0, void 0, function () {
            var _a, method, spath, args, serverName, AccountID, newActions, inputs, preHookOutData, output, _b, _c;
            return __generator(this, function (_d) {
                switch (_d.label) {
                    case 0:
                        _a = this.config, method = _a.method, spath = _a.spath, args = _a.args, serverName = _a.serverName;
                        return [4 /*yield*/, this.handleCredentials()];
                    case 1:
                        _d.sent();
                        AccountID = lodash_2.get(this.projectConfig, 'credentials.AccountID');
                        report_1.default({
                            type: 'action',
                            // TODO: 后续 可以最后一个 AccountID 删除以及sls加工数据格式。为了不影响现在的数据上报，暂时保持以前的数据格式
                            content: this.projectConfig.component + "||" + AccountID + "||" + AccountID,
                        });
                        return [4 /*yield*/, this.getNewActions({})];
                    case 2:
                        newActions = _d.sent();
                        inputs = utils_1.getInputs(this.projectConfig, {
                            method: method,
                            args: args,
                            spath: spath,
                            serverName: serverName,
                        });
                        this.hook = new hook_1.default(inputs);
                        return [4 /*yield*/, this.hook.init(newActions).executePreHook()];
                    case 3:
                        preHookOutData = _d.sent();
                        return [4 /*yield*/, this.executeCommand(preHookOutData)];
                    case 4:
                        output = _d.sent();
                        _c = (_b = this.hook).init;
                        return [4 /*yield*/, this.getNewActions({ output: output })];
                    case 5: return [4 /*yield*/, _c.apply(_b, [_d.sent()]).executeAfterHook({ output: output })];
                    case 6:
                        _d.sent();
                        return [2 /*return*/, output];
                }
            });
        });
    };
    ComponentExec.prototype.getNewActions = function (_a) {
        var output = _a.output;
        return __awaiter(this, void 0, void 0, function () {
            var _b, method, spath, serverName, globalArgs, that, tempData, realVariables, actions, useActions;
            return __generator(this, function (_c) {
                switch (_c.label) {
                    case 0:
                        _b = this.config, method = _b.method, spath = _b.spath, serverName = _b.serverName, globalArgs = _b.globalArgs;
                        that = {
                            name: serverName,
                            access: this.projectConfig.access,
                            actions: this.projectConfig.actions,
                            component: this.projectConfig.component,
                            props: this.projectConfig.props,
                            output: output,
                        };
                        tempData = { this: that };
                        this.config.parse.setCredentials(this.projectConfig.credentials);
                        return [4 /*yield*/, this.config.parse.init(tempData)];
                    case 1:
                        realVariables = (_c.sent()).realVariables;
                        this.projectConfig.actions = lodash_2.get(realVariables, ['services', serverName, 'actions']);
                        this.projectConfig.props = lodash_2.get(realVariables, ['services', serverName, 'props']);
                        actions = utils_1.getActions(this.projectConfig, {
                            method: method,
                            spath: spath,
                        });
                        useActions = (globalArgs === null || globalArgs === void 0 ? void 0 : globalArgs.skipActions) || (globalArgs === null || globalArgs === void 0 ? void 0 : globalArgs.help);
                        // actions 的解析动作包含了 config 魔法变量
                        return [2 /*return*/, useActions ? [] : actions];
                }
            });
        });
    };
    ComponentExec.prototype.executeCommand = function (payload) {
        return __awaiter(this, void 0, void 0, function () {
            var _a, method, spath, args, serverName, inputs, registry, instance, res;
            return __generator(this, function (_b) {
                switch (_b.label) {
                    case 0:
                        _a = this.config, method = _a.method, spath = _a.spath, args = _a.args, serverName = _a.serverName;
                        inputs = lodash_2.get(payload, 'type') === 'plugin'
                            ? lodash_2.get(payload, 'data')
                            : utils_1.getInputs(this.projectConfig, {
                                method: method,
                                args: args,
                                spath: spath,
                                serverName: serverName,
                                output: lodash_2.get(payload, 'data'),
                            });
                        this.debugForJest(inputs, { method: method });
                        return [4 /*yield*/, libs_1.getSetConfig('registry', constant_1.DEFAULT_REGIRSTRY)];
                    case 1:
                        registry = _b.sent();
                        return [4 /*yield*/, load_1.loadComponent(this.projectConfig.component, registry)];
                    case 2:
                        instance = _b.sent();
                        return [4 /*yield*/, this.invokeMethod(instance, inputs)];
                    case 3:
                        res = _b.sent();
                        return [2 /*return*/, typeof res === 'object' ? JSON.parse(JSON.stringify(res)) : res];
                }
            });
        });
    };
    ComponentExec.prototype.debugForJest = function (inputs, _a) {
        var method = _a.method;
        if (process.env['serverless-devs-debug'] === 'true') {
            var newInputs = lodash_1.assign({}, inputs);
            var credentials = newInputs.credentials;
            if (credentials) {
                for (var key in credentials) {
                    var val = lodash_1.toString(credentials[key]);
                    var len = val.length;
                    newInputs.credentials[key] =
                        len > 6 ? val.slice(0, 3) + '*'.repeat(len - 6) + val.slice(len - 3) : val;
                }
            }
            console.log("project:" + this.projectConfig.serviceName + " component:" + this.projectConfig.component + " method: " + method + " inputs: " + JSON.stringify(newInputs, null, 2) + " ");
        }
    };
    ComponentExec.prototype.invokeMethod = function (instance, inputs) {
        return __awaiter(this, void 0, void 0, function () {
            var _a, method, specifyService, result, error_1, result, error_2, tips;
            return __generator(this, function (_b) {
                switch (_b.label) {
                    case 0:
                        _a = this.config, method = _a.method, specifyService = _a.specifyService;
                        if (!specifyService) return [3 /*break*/, 5];
                        if (!instance[method]) return [3 /*break*/, 4];
                        _b.label = 1;
                    case 1:
                        _b.trys.push([1, 3, , 4]);
                        return [4 /*yield*/, instance[method](inputs)];
                    case 2:
                        result = _b.sent();
                        return [2 /*return*/, result];
                    case 3:
                        error_1 = _b.sent();
                        utils_1.throwError({
                            error: error_1,
                            serviceName: this.projectConfig.serviceName,
                        });
                        return [3 /*break*/, 4];
                    case 4: 
                    // 方法不存在，此时系统将会认为是未找到组件方法，系统的exit code为100；
                    throw new Error(JSON.stringify({
                        code: 100,
                        message: "The [" + method + "] command was not found.",
                        tips: "Please check the component " + this.projectConfig.component + " has the " + method + " method. Serverless Devs documents\uFF1A" + chalk_1.default.underline('https://github.com/Serverless-Devs/Serverless-Devs/blob/master/docs/zh/command'),
                    }));
                    case 5:
                        if (!instance[method]) return [3 /*break*/, 10];
                        _b.label = 6;
                    case 6:
                        _b.trys.push([6, 8, , 9]);
                        return [4 /*yield*/, instance[method](inputs)];
                    case 7:
                        result = _b.sent();
                        return [2 /*return*/, result];
                    case 8:
                        error_2 = _b.sent();
                        utils_1.throwError({
                            error: error_2,
                            serviceName: this.projectConfig.serviceName,
                        });
                        return [3 /*break*/, 9];
                    case 9: return [3 /*break*/, 11];
                    case 10:
                        tips = "Please check the component " + this.projectConfig.component + " has the " + method + " method. Serverless Devs documents\uFF1Ahttps://github.com/Serverless-Devs/Serverless-Devs/blob/master/docs/zh/command";
                        utils_1.humanWarning(tips);
                        logger_1.logger.log(chalk_1.default.grey("The [" + method + "] command was not found.\n"));
                        _b.label = 11;
                    case 11: return [2 /*return*/];
                }
            });
        });
    };
    return ComponentExec;
}());
exports.default = ComponentExec;
