// @serverless-devs/core/dist/libs/common.js
"use strict";
/**
 * @description 业务通用代码
 */
var __awaiter = (this && this.__awaiter) || function (thisArg, _arguments, P, generator) {
    function adopt(value) { return value instanceof P ? value : new P(function (resolve) { resolve(value); }); }
    return new (P || (P = Promise))(function (resolve, reject) {
        function fulfilled(value) { try { step(generator.next(value)); } catch (e) { reject(e); } }
        function rejected(value) { try { step(generator["throw"](value)); } catch (e) { reject(e); } }
        function step(result) { result.done ? resolve(result.value) : adopt(result.value).then(fulfilled, rejected); }
        step((generator = generator.apply(thisArg, _arguments || [])).next());
    });
};
var __generator = (this && this.__generator) || function (thisArg, body) {
    var _ = { label: 0, sent: function() { if (t[0] & 1) throw t[1]; return t[1]; }, trys: [], ops: [] }, f, y, t, g;
    return g = { next: verb(0), "throw": verb(1), "return": verb(2) }, typeof Symbol === "function" && (g[Symbol.iterator] = function() { return this; }), g;
    function verb(n) { return function (v) { return step([n, v]); }; }
    function step(op) {
        if (f) throw new TypeError("Generator is already executing.");
        while (_) try {
            if (f = 1, y && (t = op[0] & 2 ? y["return"] : op[0] ? y["throw"] || ((t = y["return"]) && t.call(y), 0) : y.next) && !(t = t.call(y, op[1])).done) return t;
            if (y = 0, t) op = [op[0] & 2, t.value];
            switch (op[0]) {
                case 0: case 1: t = op; break;
                case 4: _.label++; return { value: op[1], done: false };
                case 5: _.label++; y = op[1]; op = [0]; continue;
                case 7: op = _.ops.pop(); _.trys.pop(); continue;
                default:
                    if (!(t = _.trys, t = t.length > 0 && t[t.length - 1]) && (op[0] === 6 || op[0] === 2)) { _ = 0; continue; }
                    if (op[0] === 3 && (!t || (op[1] > t[0] && op[1] < t[3]))) { _.label = op[1]; break; }
                    if (op[0] === 6 && _.label < t[1]) { _.label = t[1]; t = op; break; }
                    if (t && _.label < t[2]) { _.label = t[2]; _.ops.push(op); break; }
                    if (t[2]) _.ops.pop();
                    _.trys.pop(); continue;
            }
            op = body.call(thisArg, _);
        } catch (e) { op = [6, e]; y = 0; } finally { f = t = 0; }
        if (op[0] & 5) throw op[1]; return { value: op[0] ? op[1] : void 0, done: true };
    }
};
var __importDefault = (this && this.__importDefault) || function (mod) {
    return (mod && mod.__esModule) ? mod : { "default": mod };
};
Object.defineProperty(exports, "__esModule", { value: true });
exports.getPid = exports.getCommand = exports.getSComponentPath = exports.S_CURRENT = exports.S_CURRENT_HOME = exports.isBetaS = exports.getRootHome = exports.getCliVersion = exports.setConfigYaml = exports.setConfig = exports.getSetConfig = exports.getConfig = exports.formatWorkspacePath = exports.useLocal = exports.isCiCdEnv = exports.getCicdEnv = exports.isDebugMode = void 0;
var os_1 = __importDefault(require("os"));
var path_1 = __importDefault(require("path"));
var fs_extra_1 = __importDefault(require("fs-extra"));
var minimist_1 = __importDefault(require("minimist"));
var lodash_1 = require("lodash");
var getYamlContent_1 = __importDefault(require("./getYamlContent"));
var getmac_1 = __importDefault(require("getmac"));
var js_yaml_1 = __importDefault(require("js-yaml"));
var semver = require('semver');
var USER_HOME = os_1.default.homedir();
// debug模式
var isDebugMode = function () {
    var args = minimist_1.default(process.argv.slice(2));
    return args.debug;
};
exports.isDebugMode = isDebugMode;
// s工具的家目录
function getCicdEnv() {
    if (process.env.HOME === '/kaniko' && process.env.BUILD_IMAGE_ENV === 'fc-backend') {
        return 'app_center';
    }
    for (var key in process.env) {
        if (key.startsWith('CLOUDSHELL'))
            return 'cloud_shell';
        if (key.startsWith('PIPELINE'))
            return 'yunxiao';
        if (key.startsWith('GITHUB'))
            return 'github';
        if (key.startsWith('GITLAB'))
            return 'gitlab';
        if (key.startsWith('JENKINS'))
            return 'jenkins';
    }
}
exports.getCicdEnv = getCicdEnv;
function isCiCdEnv() {
    return lodash_1.includes(['app_center', 'cloud_shell', 'yunxiao', 'github', 'gitlab', 'jenkins'], getCicdEnv());
}
exports.isCiCdEnv = isCiCdEnv;
function useLocal() {
    return process.env.default_serverless_devs_registry_mode === 'local';
}
exports.useLocal = useLocal;
function formatWorkspacePath(val) {
    return val.replace(/~/, USER_HOME);
}
exports.formatWorkspacePath = formatWorkspacePath;
function getConfig(key, defaultValue) {
    var sJsonPath = path_1.default.join(getRootHome(), 'config', 's.json');
    if (fs_extra_1.default.existsSync(sJsonPath)) {
        var data = fs_extra_1.default.readJsonSync(sJsonPath);
        var val = key ? data[key] : data;
        return val || defaultValue;
    }
}
exports.getConfig = getConfig;
function getSetConfig(key, defaultValue) {
    return __awaiter(this, void 0, void 0, function () {
        var setConfigPath, res;
        return __generator(this, function (_a) {
            switch (_a.label) {
                case 0:
                    setConfigPath = path_1.default.join(getRootHome(), 'set-config.yml');
                    return [4 /*yield*/, getYamlContent_1.default(setConfigPath)];
                case 1:
                    res = _a.sent();
                    if (!res)
                        return [2 /*return*/, defaultValue];
                    return [2 /*return*/, res[key]];
            }
        });
    });
}
exports.getSetConfig = getSetConfig;
function setConfig(key, value) {
    var _a, _b;
    if (key === 'workspace') {
        var shomedir = path_1.default.join(USER_HOME, '.s');
        var sJsonPath_1 = path_1.default.join(shomedir, 'config', 's.json');
        if (fs_extra_1.default.existsSync(sJsonPath_1)) {
            var data = fs_extra_1.default.readJsonSync(sJsonPath_1);
            data[key] = formatWorkspacePath(value);
            fs_extra_1.default.writeJsonSync(sJsonPath_1, data);
        }
        else {
            fs_extra_1.default.ensureFileSync(sJsonPath_1);
            fs_extra_1.default.writeJsonSync(sJsonPath_1, (_a = {}, _a[key] = formatWorkspacePath(value), _a));
        }
        return;
    }
    var sJsonPath = path_1.default.join(getRootHome(), 'config', 's.json');
    if (fs_extra_1.default.existsSync(sJsonPath)) {
        var data = fs_extra_1.default.readJsonSync(sJsonPath);
        data[key] = value;
        fs_extra_1.default.writeJsonSync(sJsonPath, data);
    }
    else {
        fs_extra_1.default.ensureFileSync(sJsonPath);
        fs_extra_1.default.writeJsonSync(sJsonPath, (_b = {}, _b[key] = value, _b));
    }
}
exports.setConfig = setConfig;
function setConfigYaml(key, value) {
    return __awaiter(this, void 0, void 0, function () {
        var filePath, data;
        var _a;
        return __generator(this, function (_b) {
            switch (_b.label) {
                case 0:
                    filePath = path_1.default.join(getRootHome(), 'set-config.yml');
                    return [4 /*yield*/, getYamlContent_1.default(filePath)];
                case 1:
                    data = _b.sent();
                    if (data) {
                        data[key] = value;
                        fs_extra_1.default.writeFileSync(filePath, js_yaml_1.default.dump(data));
                    }
                    else {
                        fs_extra_1.default.ensureFileSync(filePath);
                        fs_extra_1.default.writeFileSync(filePath, js_yaml_1.default.dump((_a = {}, _a[key] = value, _a)));
                    }
                    return [2 /*return*/];
            }
        });
    });
}
exports.setConfigYaml = setConfigYaml;
var getCliVersion = function (defaultValue) {
    var CLI_VERSION = process.env.CLI_VERSION;
    return CLI_VERSION || defaultValue;
};
exports.getCliVersion = getCliVersion;
function getRootHome() {
    var shomedir = path_1.default.join(USER_HOME, '.s');
    var sJsonPath = path_1.default.join(shomedir, 'config', 's.json');
    if (fs_extra_1.default.existsSync(sJsonPath)) {
        var data = fs_extra_1.default.readJsonSync(sJsonPath);
        return data.workspace ? formatWorkspacePath(data.workspace) : shomedir;
    }
    // 不存在 ～/.s/config/s.json
    if (semver.gt(exports.getCliVersion('0.0.0'), '2.0.92')) {
        var env = getCicdEnv();
        if (env === 'yunxiao')
            return path_1.default.join(USER_HOME, '.cache', '.s');
    }
    return shomedir;
}
exports.getRootHome = getRootHome;
var isBetaS = function () { return exports.getCliVersion('0.0.0').includes('beta'); };
exports.isBetaS = isBetaS;
exports.S_CURRENT_HOME = path_1.default.join(process.cwd(), '.s');
exports.S_CURRENT = path_1.default.join(process.cwd(), './');
var getSComponentPath = function () { return path_1.default.join(getRootHome(), 'components'); };
exports.getSComponentPath = getSComponentPath;
var getCommand = function () {
    try {
        var command = JSON.parse(process.env['serverless_devs_temp_argv']);
        return command ? "s " + command.join(' ') : undefined;
    }
    catch (error) { }
};
exports.getCommand = getCommand;
var getPid = function () {
    try {
        return getmac_1.default().replace(/:/g, '');
    }
    catch (error) {
        return 'unknown';
    }
};
exports.getPid = getPid;
