// @serverless-devs/core/dist/libs/utils.js
"use strict";
/**
 * @description 用于存放工具函数
 */
var __createBinding = (this && this.__createBinding) || (Object.create ? (function(o, m, k, k2) {
    if (k2 === undefined) k2 = k;
    Object.defineProperty(o, k2, { enumerable: true, get: function() { return m[k]; } });
}) : (function(o, m, k, k2) {
    if (k2 === undefined) k2 = k;
    o[k2] = m[k];
}));
var __setModuleDefault = (this && this.__setModuleDefault) || (Object.create ? (function(o, v) {
    Object.defineProperty(o, "default", { enumerable: true, value: v });
}) : function(o, v) {
    o["default"] = v;
});
var __importStar = (this && this.__importStar) || function (mod) {
    if (mod && mod.__esModule) return mod;
    var result = {};
    if (mod != null) for (var k in mod) if (k !== "default" && Object.prototype.hasOwnProperty.call(mod, k)) __createBinding(result, mod, k);
    __setModuleDefault(result, mod);
    return result;
};
var __importDefault = (this && this.__importDefault) || function (mod) {
    return (mod && mod.__esModule) ? mod : { "default": mod };
};
Object.defineProperty(exports, "__esModule", { value: true });
exports.getAvailablePort = exports.sleep = exports.writeJsonFile = exports.readJsonFile = exports.getGlobalArgs = exports.getServerlessDevsTempArgv = exports.makeUnderLine = void 0;
var fs = __importStar(require("fs-extra"));
var lodash_1 = require("lodash");
var minimist_1 = __importDefault(require("minimist"));
var chalk_1 = __importDefault(require("chalk"));
var net_1 = __importDefault(require("net"));
var makeUnderLine = function (text) {
    var matchs = text.match(/http[s]?:\/\/[^\s|,]+/);
    if (matchs) {
        return text.replace(matchs[0], chalk_1.default.underline(matchs[0]));
    }
    else {
        return text;
    }
};
exports.makeUnderLine = makeUnderLine;
function getServerlessDevsTempArgv() {
    var serverless_devs_temp_argv = process.env.serverless_devs_temp_argv;
    if (lodash_1.isEmpty(serverless_devs_temp_argv)) {
        return {};
    }
    try {
        var tempArgv = JSON.parse(serverless_devs_temp_argv);
        return getGlobalArgs(tempArgv);
    }
    catch (error) {
        return {};
    }
}
exports.getServerlessDevsTempArgv = getServerlessDevsTempArgv;
function getGlobalArgs(args) {
    if (lodash_1.isEmpty(args))
        return { _: [] };
    var newArgs = [];
    var _argsObj = [];
    var lastVal;
    var nextRemove;
    for (var index in args) {
        var val = lodash_1.trim(args[index]);
        // 将参数放到_argsObj, - 或者 -- 开头
        if (lodash_1.startsWith(val, '-') || _argsObj.length > 0) {
            _argsObj.push(val);
        }
        if (nextRemove) {
            nextRemove = false;
            continue;
        }
        if (/\s/.test(val) && lodash_1.startsWith(lastVal, '-')) {
            // 对包含空格的参数 单独处理
            newArgs.pop();
        }
        else if (lodash_1.startsWith(val, '-') && !lodash_1.startsWith(val, '--') && val.length > 2) {
            // 对类似 -la 的参数 单独处理
            nextRemove = true;
        }
        else {
            newArgs.push(val);
        }
        lastVal = val;
    }
    var data = minimist_1.default(newArgs, {
        alias: {
            template: 't',
            access: 'a',
            help: 'h',
            version: 'v',
        },
        string: ['access', 'template', 'env'],
        boolean: ['debug', 'skip-actions', 'help', 'version'],
    });
    return lodash_1.assign({ _argsObj: _argsObj }, data);
}
exports.getGlobalArgs = getGlobalArgs;
function readJsonFile(filePath) {
    if (fs.existsSync(filePath)) {
        var data = fs.readFileSync(filePath, 'utf8');
        try {
            return JSON.parse(data);
        }
        catch (error) { }
    }
}
exports.readJsonFile = readJsonFile;
function writeJsonFile(filePath, data) {
    fs.writeFileSync(filePath, JSON.stringify(data, null, 2), 'utf-8');
}
exports.writeJsonFile = writeJsonFile;
function sleep(timer) {
    return new Promise(function (resolve) {
        setTimeout(function () { return resolve(true); }, timer);
    });
}
exports.sleep = sleep;
function getAvailablePort(port) {
    if (port === void 0) { port = 3000; }
    var server = net_1.default.createServer().listen(port);
    return new Promise(function (resolve, reject) {
        server.on('listening', function () {
            server.close();
            resolve(port);
        });
        server.on('error', function (err) {
            if (err.code === 'EADDRINUSE') {
                resolve(getAvailablePort(port + 1));
            }
            else {
                reject(err);
            }
        });
    });
}
exports.getAvailablePort = getAvailablePort;
