// @serverless-devs/core/dist/common/credential/index.js
"use strict";
var __importDefault = (this && this.__importDefault) || function (mod) {
    return (mod && mod.__esModule) ? mod : { "default": mod };
};
Object.defineProperty(exports, "__esModule", { value: true });
exports.getAccountId = exports.setKnownCredential = exports.setCredential = exports.getCredentialAliasList = exports.getCredentialFromEnv = exports.decryptCredential = exports.getCredential = void 0;
var getCredential_1 = require("./getCredential");
Object.defineProperty(exports, "getCredential", { enumerable: true, get: function () { return __importDefault(getCredential_1).default; } });
Object.defineProperty(exports, "decryptCredential", { enumerable: true, get: function () { return getCredential_1.decryptCredential; } });
Object.defineProperty(exports, "getCredentialFromEnv", { enumerable: true, get: function () { return getCredential_1.getCredentialFromEnv; } });
Object.defineProperty(exports, "getCredentialAliasList", { enumerable: true, get: function () { return getCredential_1.getCredentialAliasList; } });
var setCredential_1 = require("./setCredential");
Object.defineProperty(exports, "setCredential", { enumerable: true, get: function () { return __importDefault(setCredential_1).default; } });
Object.defineProperty(exports, "setKnownCredential", { enumerable: true, get: function () { return setCredential_1.setKnownCredential; } });
var getAccountId_1 = require("./getAccountId");
Object.defineProperty(exports, "getAccountId", { enumerable: true, get: function () { return __importDefault(getAccountId_1).default; } });
