// @serverless-devs/core/dist/common/credential/setCredential.js
"use strict";
var __assign = (this && this.__assign) || function () {
    __assign = Object.assign || function(t) {
        for (var s, i = 1, n = arguments.length; i < n; i++) {
            s = arguments[i];
            for (var p in s) if (Object.prototype.hasOwnProperty.call(s, p))
                t[p] = s[p];
        }
        return t;
    };
    return __assign.apply(this, arguments);
};
var __awaiter = (this && this.__awaiter) || function (thisArg, _arguments, P, generator) {
    function adopt(value) { return value instanceof P ? value : new P(function (resolve) { resolve(value); }); }
    return new (P || (P = Promise))(function (resolve, reject) {
        function fulfilled(value) { try { step(generator.next(value)); } catch (e) { reject(e); } }
        function rejected(value) { try { step(generator["throw"](value)); } catch (e) { reject(e); } }
        function step(result) { result.done ? resolve(result.value) : adopt(result.value).then(fulfilled, rejected); }
        step((generator = generator.apply(thisArg, _arguments || [])).next());
    });
};
var __generator = (this && this.__generator) || function (thisArg, body) {
    var _ = { label: 0, sent: function() { if (t[0] & 1) throw t[1]; return t[1]; }, trys: [], ops: [] }, f, y, t, g;
    return g = { next: verb(0), "throw": verb(1), "return": verb(2) }, typeof Symbol === "function" && (g[Symbol.iterator] = function() { return this; }), g;
    function verb(n) { return function (v) { return step([n, v]); }; }
    function step(op) {
        if (f) throw new TypeError("Generator is already executing.");
        while (_) try {
            if (f = 1, y && (t = op[0] & 2 ? y["return"] : op[0] ? y["throw"] || ((t = y["return"]) && t.call(y), 0) : y.next) && !(t = t.call(y, op[1])).done) return t;
            if (y = 0, t) op = [op[0] & 2, t.value];
            switch (op[0]) {
                case 0: case 1: t = op; break;
                case 4: _.label++; return { value: op[1], done: false };
                case 5: _.label++; y = op[1]; op = [0]; continue;
                case 7: op = _.ops.pop(); _.trys.pop(); continue;
                default:
                    if (!(t = _.trys, t = t.length > 0 && t[t.length - 1]) && (op[0] === 6 || op[0] === 2)) { _ = 0; continue; }
                    if (op[0] === 3 && (!t || (op[1] > t[0] && op[1] < t[3]))) { _.label = op[1]; break; }
                    if (op[0] === 6 && _.label < t[1]) { _.label = t[1]; t = op; break; }
                    if (t && _.label < t[2]) { _.label = t[2]; _.ops.push(op); break; }
                    if (t[2]) _.ops.pop();
                    _.trys.pop(); continue;
            }
            op = body.call(thisArg, _);
        } catch (e) { op = [6, e]; y = 0; } finally { f = t = 0; }
        if (op[0] & 5) throw op[1]; return { value: op[0] ? op[1] : void 0, done: true };
    }
};
var __importDefault = (this && this.__importDefault) || function (mod) {
    return (mod && mod.__esModule) ? mod : { "default": mod };
};
Object.defineProperty(exports, "__esModule", { value: true });
exports.setKnownCredential = void 0;
var inquirer_1 = __importDefault(require("inquirer"));
var fs_extra_1 = __importDefault(require("fs-extra"));
var path_1 = __importDefault(require("path"));
var js_yaml_1 = __importDefault(require("js-yaml"));
var constant_1 = require("./constant");
var libs_1 = require("../../libs");
var logger_1 = require("../../logger");
var getAccountId_1 = __importDefault(require("./getAccountId"));
var spinner_1 = __importDefault(require("../spinner"));
var lodash_1 = require("lodash");
var chalk_1 = __importDefault(require("chalk"));
var Crypto = require('crypto-js');
function handleCustom(info) {
    return __awaiter(this, void 0, void 0, function () {
        var option, name, _a, key, value;
        return __generator(this, function (_b) {
            switch (_b.label) {
                case 0:
                    option = [
                        {
                            type: 'list',
                            name: 'name',
                            message: 'Please select a type:',
                            choices: [
                                { name: 'Add key-value pairs', value: 'add' },
                                { name: 'End of adding key-value pairs', value: 'over' },
                            ],
                        },
                    ];
                    return [4 /*yield*/, inquirer_1.default.prompt(option)];
                case 1:
                    name = (_b.sent()).name;
                    if (!(name === 'add')) return [3 /*break*/, 4];
                    return [4 /*yield*/, inquirer_1.default.prompt([
                            {
                                type: 'input',
                                message: 'Please enter key: ',
                                name: 'key',
                            },
                            {
                                type: 'input',
                                message: 'Please enter value: ',
                                name: 'value',
                            },
                        ])];
                case 2:
                    _a = _b.sent(), key = _a.key, value = _a.value;
                    info[key] = value;
                    return [4 /*yield*/, handleCustom(info)];
                case 3:
                    _b.sent();
                    _b.label = 4;
                case 4: return [2 /*return*/];
            }
        });
    });
}
function secret(tempAccess) {
    if (lodash_1.isEmpty(tempAccess))
        return tempAccess;
    var tempSecretAccess = {};
    for (var eveValue in tempAccess) {
        var valueLength = tempAccess[eveValue].length;
        tempSecretAccess[eveValue] =
            valueLength > 6
                ? tempAccess[eveValue].slice(0, 3) +
                    '*'.repeat(valueLength - 6) +
                    tempAccess[eveValue].slice(valueLength - 3, valueLength)
                : tempAccess[eveValue];
    }
    return tempSecretAccess;
}
function output(_a) {
    var info = _a.info, accessAlias = _a.accessAlias;
    console.log('');
    logger_1.logger.output(__assign({ Alias: accessAlias }, secret(info)), 2);
    console.log('');
    spinner_1.default('Configuration successful').succeed();
}
function encrypt(info) {
    if (info === void 0) { info = {}; }
    var cloneInfo = Object.assign({}, info);
    Object.keys(cloneInfo).forEach(function (key) {
        var ciphertext = Crypto.AES.encrypt(cloneInfo[key], 'SecretKey123');
        cloneInfo[key] = ciphertext.toString();
    });
    return cloneInfo;
}
function writeData(data) {
    return __awaiter(this, void 0, void 0, function () {
        var info, accessAlias, filePath, content, providerAliasKeys, tempArgv, option, name_1, accessAliasObj, aliasName;
        var _a, _b;
        return __generator(this, function (_c) {
            switch (_c.label) {
                case 0:
                    info = data.info, accessAlias = data.accessAlias;
                    filePath = path_1.default.join(libs_1.getRootHome(), 'access.yaml');
                    return [4 /*yield*/, libs_1.getYamlContent(filePath)];
                case 1:
                    content = _c.sent();
                    if (!content) return [3 /*break*/, 9];
                    providerAliasKeys = Object.keys(content);
                    tempArgv = libs_1.getServerlessDevsTempArgv();
                    if (!(providerAliasKeys.includes(accessAlias) && !tempArgv.f)) return [3 /*break*/, 7];
                    option = [
                        {
                            type: 'list',
                            name: 'name',
                            message: 'Alias already exists. Please select a type:',
                            choices: [
                                { name: 'overwrite', value: 'overwrite' },
                                { name: 'rename', value: 'rename' },
                                { name: 'exit', value: 'exit' },
                            ],
                        },
                    ];
                    return [4 /*yield*/, inquirer_1.default.prompt(option)];
                case 2:
                    name_1 = (_c.sent()).name;
                    if (name_1 === 'overwrite') {
                        content[accessAlias] = encrypt(info);
                        fs_extra_1.default.writeFileSync(filePath, js_yaml_1.default.dump(content));
                        output({ info: info, accessAlias: accessAlias });
                    }
                    if (!(name_1 === 'rename')) return [3 /*break*/, 6];
                    _a = {
                        type: 'input',
                        message: 'Please create alias for key pair. If not, please enter to skip',
                        name: 'aliasName'
                    };
                    return [4 /*yield*/, getAlias()];
                case 3:
                    accessAliasObj = [
                        (_a.default = _c.sent(),
                            _a)
                    ];
                    return [4 /*yield*/, inquirer_1.default.prompt(accessAliasObj)];
                case 4:
                    aliasName = (_c.sent()).aliasName;
                    return [4 /*yield*/, writeData({ info: info, accessAlias: aliasName })];
                case 5: return [2 /*return*/, _c.sent()];
                case 6: return [3 /*break*/, 8];
                case 7:
                    try {
                        content[accessAlias] = encrypt(info);
                        fs_extra_1.default.writeFileSync(filePath, js_yaml_1.default.dump(content));
                        output({ info: info, accessAlias: accessAlias });
                    }
                    catch (err) {
                        throw new Error('Configuration failed');
                    }
                    _c.label = 8;
                case 8: return [3 /*break*/, 10];
                case 9:
                    try {
                        fs_extra_1.default.ensureFileSync(filePath);
                        fs_extra_1.default.writeFileSync(filePath, js_yaml_1.default.dump((_b = {}, _b[accessAlias] = encrypt(info), _b)));
                        output({ info: info, accessAlias: accessAlias });
                    }
                    catch (err) {
                        throw new Error('Configuration failed');
                    }
                    _c.label = 10;
                case 10: return [2 /*return*/, data];
            }
        });
    });
}
function getAlias() {
    return __awaiter(this, void 0, void 0, function () {
        var filePath, info, keys, max_1;
        return __generator(this, function (_a) {
            switch (_a.label) {
                case 0:
                    filePath = path_1.default.join(libs_1.getRootHome(), 'access.yaml');
                    if (!fs_extra_1.default.existsSync(filePath)) return [3 /*break*/, 2];
                    return [4 /*yield*/, libs_1.getYamlContent(filePath)];
                case 1:
                    info = _a.sent();
                    keys = info ? Object.keys(info).filter(function (item) { return item.startsWith('default'); }) : [];
                    if (keys.length === 0) {
                        return [2 /*return*/, 'default'];
                    }
                    max_1 = '0';
                    keys.forEach(function (item) {
                        var _a = item.split('-'), end = _a[1];
                        if (!isNaN(parseInt(end)) && end > max_1) {
                            max_1 = end;
                        }
                    });
                    return [2 /*return*/, "default-" + (parseInt(max_1) + 1)];
                case 2: return [2 /*return*/, 'default'];
            }
        });
    });
}
/**
 * @param args 可选参数，接收设置密钥的key，如果不传方法内部提供了设置密钥的相关模版
 */
function setCredential() {
    var args = [];
    for (var _i = 0; _i < arguments.length; _i++) {
        args[_i] = arguments[_i];
    }
    return __awaiter(this, void 0, void 0, function () {
        var selectedProvider, info, answers, accessAlias, accessAliasObj, res, argsPrompt, promptList, data, error_1, Alias;
        var _a;
        return __generator(this, function (_b) {
            switch (_b.label) {
                case 0:
                    if (!(args.length > 0)) return [3 /*break*/, 1];
                    selectedProvider = 'params';
                    return [3 /*break*/, 3];
                case 1: return [4 /*yield*/, inquirer_1.default.prompt(constant_1.checkProviderList)];
                case 2:
                    answers = _b.sent();
                    selectedProvider = answers.provider;
                    _b.label = 3;
                case 3:
                    if (['alibaba', 'aws', 'azure', 'baidu', 'google', 'huawei', 'tencent'].includes(selectedProvider)) {
                        console.log("\uD83E\uDDED Refer to the document for " + selectedProvider + " key: ", {
                            alibaba: 'http://config.devsapp.net/account/alibaba',
                            aws: 'http://config.devsapp.net/account/aws',
                            azure: 'http://config.devsapp.net/account/azure',
                            baidu: 'http://config.devsapp.net/account/baidu',
                            google: 'http://config.devsapp.net/account/gcp',
                            huawei: 'http://config.devsapp.net/account/huawei',
                            tencent: 'http://config.devsapp.net/account/tencent',
                        }[selectedProvider]);
                    }
                    _a = {
                        type: 'input',
                        message: 'Please create alias for key pair. If not, please enter to skip',
                        name: 'aliasName'
                    };
                    return [4 /*yield*/, getAlias()];
                case 4:
                    accessAliasObj = (_a.default = _b.sent(),
                        _a);
                    if (!(selectedProvider === 'custom')) return [3 /*break*/, 7];
                    return [4 /*yield*/, handleCustom((info = {}))];
                case 5:
                    _b.sent();
                    return [4 /*yield*/, inquirer_1.default.prompt([accessAliasObj])];
                case 6:
                    res = _b.sent();
                    accessAlias = res.aliasName;
                    return [3 /*break*/, 13];
                case 7:
                    argsPrompt = args.map(function (item) { return ({
                        type: 'input',
                        message: item,
                        name: item,
                    }); });
                    promptList = selectedProvider === 'params' ? argsPrompt : constant_1.providerCollection[selectedProvider];
                    promptList.push(accessAliasObj);
                    return [4 /*yield*/, inquirer_1.default.prompt(promptList)];
                case 8:
                    info = _b.sent();
                    if (!(selectedProvider === 'alibaba')) return [3 /*break*/, 12];
                    _b.label = 9;
                case 9:
                    _b.trys.push([9, 11, , 12]);
                    return [4 /*yield*/, getAccountId_1.default(info)];
                case 10:
                    data = _b.sent();
                    info.AccountID = data.AccountId;
                    return [3 /*break*/, 12];
                case 11:
                    error_1 = _b.sent();
                    throw new Error(JSON.stringify({
                        message: 'You are configuring an incorrect Alibaba Cloud SecretKey.',
                        tips: "Please check the accuracy of Alibaba Cloud SecretKey. documents: " + chalk_1.default.underline('https://github.com/Serverless-Devs/Serverless-Devs/blob/master/docs/zh/default_provider_config/alibabacloud.md'),
                    }));
                case 12:
                    Object.keys(info).forEach(function (item) {
                        if (item === 'aliasName') {
                            accessAlias = info[item];
                            delete info[item];
                        }
                    });
                    _b.label = 13;
                case 13: return [4 /*yield*/, writeData({ info: info, accessAlias: accessAlias })];
                case 14:
                    Alias = (_b.sent()).accessAlias;
                    return [2 /*return*/, __assign({ Alias: Alias }, info)];
            }
        });
    });
}
/**
 * @param args 接收设置密钥的key
 */
function setKnownCredential(info, accessAlias) {
    return __awaiter(this, void 0, void 0, function () {
        var aliasName, _a;
        return __generator(this, function (_b) {
            switch (_b.label) {
                case 0:
                    _a = accessAlias;
                    if (_a) return [3 /*break*/, 2];
                    return [4 /*yield*/, getAlias()];
                case 1:
                    _a = (_b.sent());
                    _b.label = 2;
                case 2:
                    aliasName = _a;
                    return [4 /*yield*/, writeData({ info: info, accessAlias: aliasName })];
                case 3:
                    _b.sent();
                    return [2 /*return*/, __assign({ Alias: aliasName }, info)];
            }
        });
    });
}
exports.setKnownCredential = setKnownCredential;
exports.default = setCredential;
