// @serverless-devs/core/dist/common/execCommand/analysis.js
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
var Analysis = /** @class */ (function () {
    function Analysis(parsedObj, dependenciesMap) {
        this.parsedObj = parsedObj;
        this.dependenciesMap = dependenciesMap;
        this.componentOrderKeyMap = {};
    }
    Analysis.prototype.getComponentOrderKeyMap = function (projKeys) {
        var _this = this;
        projKeys.forEach(function (key) {
            _this.componentOrderKeyMap[key] = 1;
        });
    };
    Analysis.prototype.calculateOrderNumber = function (componentKeys) {
        var _this = this;
        componentKeys.forEach(function (key) {
            var dependenciesInstance = _this.dependenciesMap[key];
            Object.keys(dependenciesInstance).forEach(function (_key) {
                if (_this.componentOrderKeyMap[_key]) {
                    _this.componentOrderKeyMap[_key] += dependenciesInstance[_key];
                }
            });
        });
    };
    Analysis.prototype.getProjectOrder = function () {
        var _this = this;
        var componentKeys = Object.keys(this.dependenciesMap).filter(function (key) { return key !== 'Global'; });
        this.getComponentOrderKeyMap(componentKeys); //
        this.calculateOrderNumber(componentKeys);
        var projectArray = Object.keys(this.componentOrderKeyMap).map(function (key) {
            return { name: key, order: _this.componentOrderKeyMap[key] };
        });
        return projectArray
            .sort(function (item1, item2) {
            return item2.order - item1.order;
        })
            .map(function (item) { return item.name; });
    };
    return Analysis;
}());
exports.default = Analysis;
