// @serverless-devs/core/dist/common/execCommand/getTemplatePath.js
"use strict";
var __awaiter = (this && this.__awaiter) || function (thisArg, _arguments, P, generator) {
    function adopt(value) { return value instanceof P ? value : new P(function (resolve) { resolve(value); }); }
    return new (P || (P = Promise))(function (resolve, reject) {
        function fulfilled(value) { try { step(generator.next(value)); } catch (e) { reject(e); } }
        function rejected(value) { try { step(generator["throw"](value)); } catch (e) { reject(e); } }
        function step(result) { result.done ? resolve(result.value) : adopt(result.value).then(fulfilled, rejected); }
        step((generator = generator.apply(thisArg, _arguments || [])).next());
    });
};
var __generator = (this && this.__generator) || function (thisArg, body) {
    var _ = { label: 0, sent: function() { if (t[0] & 1) throw t[1]; return t[1]; }, trys: [], ops: [] }, f, y, t, g;
    return g = { next: verb(0), "throw": verb(1), "return": verb(2) }, typeof Symbol === "function" && (g[Symbol.iterator] = function() { return this; }), g;
    function verb(n) { return function (v) { return step([n, v]); }; }
    function step(op) {
        if (f) throw new TypeError("Generator is already executing.");
        while (_) try {
            if (f = 1, y && (t = op[0] & 2 ? y["return"] : op[0] ? y["throw"] || ((t = y["return"]) && t.call(y), 0) : y.next) && !(t = t.call(y, op[1])).done) return t;
            if (y = 0, t) op = [op[0] & 2, t.value];
            switch (op[0]) {
                case 0: case 1: t = op; break;
                case 4: _.label++; return { value: op[1], done: false };
                case 5: _.label++; y = op[1]; op = [0]; continue;
                case 7: op = _.ops.pop(); _.trys.pop(); continue;
                default:
                    if (!(t = _.trys, t = t.length > 0 && t[t.length - 1]) && (op[0] === 6 || op[0] === 2)) { _ = 0; continue; }
                    if (op[0] === 3 && (!t || (op[1] > t[0] && op[1] < t[3]))) { _.label = op[1]; break; }
                    if (op[0] === 6 && _.label < t[1]) { _.label = t[1]; t = op; break; }
                    if (t && _.label < t[2]) { _.label = t[2]; _.ops.push(op); break; }
                    if (t[2]) _.ops.pop();
                    _.trys.pop(); continue;
            }
            op = body.call(thisArg, _);
        } catch (e) { op = [6, e]; y = 0; } finally { f = t = 0; }
        if (op[0] & 5) throw op[1]; return { value: op[0] ? op[1] : void 0, done: true };
    }
};
var __importDefault = (this && this.__importDefault) || function (mod) {
    return (mod && mod.__esModule) ? mod : { "default": mod };
};
Object.defineProperty(exports, "__esModule", { value: true });
exports.getTemplatePath = exports.transforYamlPath = void 0;
var path_1 = __importDefault(require("path"));
var fs_extra_1 = __importDefault(require("fs-extra"));
var libs_1 = require("../../libs");
var lodash_1 = require("lodash");
var js_yaml_1 = __importDefault(require("js-yaml"));
var chalk_1 = __importDefault(require("chalk"));
var extend2_1 = __importDefault(require("extend2"));
var parseYaml_1 = __importDefault(require("../parseYaml"));
function checkYaml(spath) {
    return __awaiter(this, void 0, void 0, function () {
        var filename, data, services, key, ele;
        return __generator(this, function (_a) {
            switch (_a.label) {
                case 0:
                    filename = path_1.default.basename(spath);
                    return [4 /*yield*/, libs_1.getYamlContent(spath)];
                case 1:
                    data = _a.sent();
                    // 校验 edition 字段
                    if (!['1.0.0', '2.0.0'].includes(lodash_1.get(data, 'edition'))) {
                        throw new Error(JSON.stringify({
                            message: "The edition field in the " + filename + " file is incorrect.",
                            tips: "Please check the edition field of " + filename + ", you can specify it as 1.0.0 or 2.0.0.",
                        }));
                    }
                    services = lodash_1.get(data, 'services');
                    // 校验 services 字段
                    if (!lodash_1.isPlainObject(services)) {
                        throw new Error(JSON.stringify({
                            message: "The services field in the " + filename + " file is incorrect.",
                            tips: "Please check the services field of " + filename + ", documents: " + chalk_1.default.underline('https://docs.serverless-devs.com/fc/yaml/readme') + ".",
                        }));
                    }
                    // 校验 component 字段
                    for (key in services) {
                        ele = services[key];
                        if (lodash_1.isEmpty(ele.component)) {
                            throw new Error(JSON.stringify({
                                message: "The component field in " + key + " service is incorrect.",
                                tips: "Please check the component field in " + key + " service, documents: " + chalk_1.default.underline('https://docs.serverless-devs.com/fc/yaml/readme') + ".",
                            }));
                        }
                    }
                    return [2 /*return*/, spath];
            }
        });
    });
}
function setupEnv(templateFile) {
    return __awaiter(this, void 0, void 0, function () {
        var spath, envPath, data, services, key, element, codeUri, codeUriEnvPath;
        return __generator(this, function (_a) {
            switch (_a.label) {
                case 0:
                    spath = path_1.default.dirname(templateFile);
                    envPath = path_1.default.join(spath, '.env');
                    if (fs_extra_1.default.existsSync(envPath)) {
                        require('dotenv').config({ path: envPath });
                    }
                    return [4 /*yield*/, libs_1.getYamlContent(templateFile)];
                case 1:
                    data = _a.sent();
                    services = lodash_1.get(data, 'services');
                    if (lodash_1.isEmpty(services))
                        return [2 /*return*/];
                    for (key in services) {
                        element = services[key];
                        codeUri = lodash_1.get(element, 'props.function.codeUri');
                        if (codeUri) {
                            codeUri = path_1.default.isAbsolute(codeUri) ? codeUri : path_1.default.join(spath, codeUri);
                            codeUriEnvPath = path_1.default.join(codeUri, '.env');
                            fs_extra_1.default.existsSync(codeUriEnvPath) && require('dotenv').config({ path: codeUriEnvPath });
                        }
                    }
                    return [2 /*return*/];
            }
        });
    });
}
function extendsYaml(spath, dotspath, data) {
    return __awaiter(this, void 0, void 0, function () {
        var extendsPath, yamlPath, dirname, fixedPath, doc, newData;
        return __generator(this, function (_a) {
            switch (_a.label) {
                case 0:
                    extendsPath = (data === null || data === void 0 ? void 0 : data.extends) ? lodash_1.first(data.extends) : undefined;
                    yamlPath = (data === null || data === void 0 ? void 0 : data.extend) ? data.extend : extendsPath;
                    if (lodash_1.isEmpty(yamlPath))
                        return [2 /*return*/];
                    if (!path_1.default.isAbsolute(yamlPath)) {
                        dirname = path_1.default.dirname(spath);
                        fixedPath = path_1.default.resolve(dirname, yamlPath);
                        if (fs_extra_1.default.existsSync(fixedPath))
                            yamlPath = fixedPath;
                    }
                    return [4 /*yield*/, libs_1.isYamlFile(yamlPath)];
                case 1:
                    _a.sent();
                    if (!(data === null || data === void 0 ? void 0 : data.vars)) return [3 /*break*/, 4];
                    return [4 /*yield*/, libs_1.getYamlContent(yamlPath)];
                case 2:
                    doc = _a.sent();
                    newData = extend2_1.default(true, doc, { vars: data.vars });
                    fs_extra_1.default.writeFileSync(dotspath, js_yaml_1.default.dump(newData));
                    return [4 /*yield*/, parseYaml_1.default(fs_extra_1.default.readFileSync(dotspath, 'utf-8'))];
                case 3: return [2 /*return*/, _a.sent()];
                case 4: return [4 /*yield*/, parseYaml_1.default(fs_extra_1.default.readFileSync(yamlPath, 'utf-8'))];
                case 5: return [2 /*return*/, _a.sent()];
            }
        });
    });
}
function transforYamlPath(spath) {
    if (spath === void 0) { spath = ''; }
    return __awaiter(this, void 0, void 0, function () {
        var data, dotspath, tmp, extend2Data;
        return __generator(this, function (_a) {
            switch (_a.label) {
                case 0: return [4 /*yield*/, libs_1.isYamlFile(spath)];
                case 1:
                    _a.sent();
                    return [4 /*yield*/, setupEnv(spath)];
                case 2:
                    _a.sent();
                    return [4 /*yield*/, parseYaml_1.default(fs_extra_1.default.readFileSync(spath, 'utf-8'))];
                case 3:
                    data = _a.sent();
                    // 兼容 extends 只取第一个即可
                    if (lodash_1.isEmpty(data === null || data === void 0 ? void 0 : data.extends) && lodash_1.isEmpty(data === null || data === void 0 ? void 0 : data.extend)) {
                        return [2 /*return*/, checkYaml(spath)];
                    }
                    dotspath = path_1.default.join(path_1.default.dirname(spath), '.s', path_1.default.basename(spath));
                    fs_extra_1.default.ensureFileSync(dotspath);
                    return [4 /*yield*/, extendsYaml(spath, dotspath, data)];
                case 4:
                    tmp = _a.sent();
                    extend2Data = extend2_1.default(true, tmp, lodash_1.omit(data, ['extends', 'extend']));
                    fs_extra_1.default.writeFileSync(dotspath, js_yaml_1.default.dump(extend2Data));
                    return [2 /*return*/, checkYaml(dotspath)];
            }
        });
    });
}
exports.transforYamlPath = transforYamlPath;
function getTemplatePath(spath) {
    if (spath === void 0) { spath = ''; }
    return __awaiter(this, void 0, void 0, function () {
        var cwd, sYamlPath, sYmlPath;
        return __generator(this, function (_a) {
            if (fs_extra_1.default.existsSync(spath)) {
                return [2 /*return*/, path_1.default.isAbsolute(spath) ? spath : path_1.default.resolve(spath)];
            }
            cwd = process.cwd();
            sYamlPath = path_1.default.join(cwd, 's.yaml');
            if (fs_extra_1.default.existsSync(sYamlPath))
                return [2 /*return*/, sYamlPath];
            sYmlPath = path_1.default.join(cwd, 's.yml');
            if (fs_extra_1.default.existsSync(sYmlPath))
                return [2 /*return*/, sYmlPath];
            throw new Error(JSON.stringify({
                message: 'the s.yaml/s.yml file was not found.',
                tips: 'Please check if the s.yaml/s.yml file exists, you can also specify it with -t.',
            }));
        });
    });
}
exports.getTemplatePath = getTemplatePath;
