// @serverless-devs/core/dist/common/execCommand/index.js
"use strict";
var __awaiter = (this && this.__awaiter) || function (thisArg, _arguments, P, generator) {
    function adopt(value) { return value instanceof P ? value : new P(function (resolve) { resolve(value); }); }
    return new (P || (P = Promise))(function (resolve, reject) {
        function fulfilled(value) { try { step(generator.next(value)); } catch (e) { reject(e); } }
        function rejected(value) { try { step(generator["throw"](value)); } catch (e) { reject(e); } }
        function step(result) { result.done ? resolve(result.value) : adopt(result.value).then(fulfilled, rejected); }
        step((generator = generator.apply(thisArg, _arguments || [])).next());
    });
};
var __generator = (this && this.__generator) || function (thisArg, body) {
    var _ = { label: 0, sent: function() { if (t[0] & 1) throw t[1]; return t[1]; }, trys: [], ops: [] }, f, y, t, g;
    return g = { next: verb(0), "throw": verb(1), "return": verb(2) }, typeof Symbol === "function" && (g[Symbol.iterator] = function() { return this; }), g;
    function verb(n) { return function (v) { return step([n, v]); }; }
    function step(op) {
        if (f) throw new TypeError("Generator is already executing.");
        while (_) try {
            if (f = 1, y && (t = op[0] & 2 ? y["return"] : op[0] ? y["throw"] || ((t = y["return"]) && t.call(y), 0) : y.next) && !(t = t.call(y, op[1])).done) return t;
            if (y = 0, t) op = [op[0] & 2, t.value];
            switch (op[0]) {
                case 0: case 1: t = op; break;
                case 4: _.label++; return { value: op[1], done: false };
                case 5: _.label++; y = op[1]; op = [0]; continue;
                case 7: op = _.ops.pop(); _.trys.pop(); continue;
                default:
                    if (!(t = _.trys, t = t.length > 0 && t[t.length - 1]) && (op[0] === 6 || op[0] === 2)) { _ = 0; continue; }
                    if (op[0] === 3 && (!t || (op[1] > t[0] && op[1] < t[3]))) { _.label = op[1]; break; }
                    if (op[0] === 6 && _.label < t[1]) { _.label = t[1]; t = op; break; }
                    if (t && _.label < t[2]) { _.label = t[2]; _.ops.push(op); break; }
                    if (t[2]) _.ops.pop();
                    _.trys.pop(); continue;
            }
            op = body.call(thisArg, _);
        } catch (e) { op = [6, e]; y = 0; } finally { f = t = 0; }
        if (op[0] & 5) throw op[1]; return { value: op[0] ? op[1] : void 0, done: true };
    }
};
var __importDefault = (this && this.__importDefault) || function (mod) {
    return (mod && mod.__esModule) ? mod : { "default": mod };
};
Object.defineProperty(exports, "__esModule", { value: true });
exports.transforYamlPath = exports.getTemplatePath = exports.execCommand = void 0;
var parse_1 = __importDefault(require("./parse"));
var lodash_1 = require("lodash");
var logger_1 = require("../../logger");
var analysis_1 = __importDefault(require("./analysis"));
var utils_1 = require("./utils");
var getTemplatePath_1 = require("./getTemplatePath");
Object.defineProperty(exports, "getTemplatePath", { enumerable: true, get: function () { return getTemplatePath_1.getTemplatePath; } });
Object.defineProperty(exports, "transforYamlPath", { enumerable: true, get: function () { return getTemplatePath_1.transforYamlPath; } });
var component_1 = __importDefault(require("./component"));
var execDaemon_1 = require("../../execDaemon");
var libs_1 = require("../../libs");
var ExecCommand = /** @class */ (function () {
    function ExecCommand(configs) {
        this.configs = configs;
        var env = configs.env;
        if (lodash_1.isPlainObject(env)) {
            for (var key in env) {
                process.env[key] = env[key];
            }
        }
        // 写入日志的时候，先确保创建了日志文件
        logger_1.makeLogFile();
    }
    ExecCommand.prototype.init = function () {
        return __awaiter(this, void 0, void 0, function () {
            var _a, syaml, serverName, _b, globalArgs, originSpath, spath, parsedObj, analysis, executeOrderList, tempCustomerCommandName;
            return __generator(this, function (_c) {
                switch (_c.label) {
                    case 0:
                        _a = this.configs, syaml = _a.syaml, serverName = _a.serverName, _b = _a.globalArgs, globalArgs = _b === void 0 ? {} : _b;
                        return [4 /*yield*/, getTemplatePath_1.getTemplatePath(syaml)];
                    case 1:
                        originSpath = _c.sent();
                        return [4 /*yield*/, getTemplatePath_1.transforYamlPath(originSpath)];
                    case 2:
                        spath = _c.sent();
                        this.parse = new parse_1.default(spath);
                        return [4 /*yield*/, this.parse.init()];
                    case 3:
                        parsedObj = _c.sent();
                        return [4 /*yield*/, this.parse.init(parsedObj.realVariables)];
                    case 4:
                        // 兼容vars下的魔法变量，需再次解析
                        parsedObj = _c.sent();
                        return [4 /*yield*/, this.warnEnvironmentVariables(parsedObj.realVariables)];
                    case 5:
                        _c.sent();
                        analysis = new analysis_1.default(parsedObj.realVariables, parsedObj.dependenciesMap);
                        executeOrderList = analysis.getProjectOrder();
                        // 私有化部署不在进行上报数据
                        if (!libs_1.useLocal()) {
                            execDaemon_1.execDaemon('reportTracker.js', { syaml: spath, access: globalArgs.access });
                        }
                        if (!(executeOrderList.length === 1 || serverName)) return [3 /*break*/, 7];
                        tempCustomerCommandName = executeOrderList[0];
                        return [4 /*yield*/, this.serviceOnlyOne({
                                realVariables: parsedObj.realVariables,
                                serverName: serverName || tempCustomerCommandName,
                                spath: originSpath,
                                specifyService: Boolean(serverName),
                            })];
                    case 6: return [2 /*return*/, _c.sent()];
                    case 7: return [4 /*yield*/, this.serviceWithMany({ executeOrderList: executeOrderList, spath: originSpath })];
                    case 8: return [2 /*return*/, _c.sent()];
                }
            });
        });
    };
    ExecCommand.prototype.serviceOnlyOne = function (_a) {
        var realVariables = _a.realVariables, serverName = _a.serverName, spath = _a.spath, specifyService = _a.specifyService;
        return __awaiter(this, void 0, void 0, function () {
            var _b, method, args, globalArgs, projectConfig, outPutData, result;
            var _c;
            return __generator(this, function (_d) {
                switch (_d.label) {
                    case 0:
                        _b = this.configs, method = _b.method, args = _b.args, globalArgs = _b.globalArgs;
                        projectConfig = utils_1.getProjectConfig(realVariables, serverName, globalArgs);
                        return [4 /*yield*/, new component_1.default({
                                projectConfig: projectConfig,
                                method: method,
                                args: args,
                                spath: spath,
                                serverName: serverName,
                                globalArgs: globalArgs,
                                specifyService: specifyService,
                                parse: this.parse,
                            }).init()];
                    case 1:
                        outPutData = _d.sent();
                        result = (_c = {}, _c[serverName] = outPutData, _c);
                        lodash_1.keys(outPutData).length === 0
                            ? logger_1.logger.log("End of method: " + method, 'green')
                            : logger_1.logger.output(result);
                        return [2 /*return*/, result];
                }
            });
        });
    };
    ExecCommand.prototype.serviceWithMany = function (_a) {
        var executeOrderList = _a.executeOrderList, spath = _a.spath;
        return __awaiter(this, void 0, void 0, function () {
            var _b, method, args, globalArgs, result, tempData, _i, executeOrderList_1, serverName, parsedObj, projectConfig, outputData;
            return __generator(this, function (_c) {
                switch (_c.label) {
                    case 0:
                        _b = this.configs, method = _b.method, args = _b.args, globalArgs = _b.globalArgs;
                        logger_1.logger.info("It is detected that your project has the following projects < " + executeOrderList.join(',') + " > to be execute");
                        result = {};
                        tempData = { services: {} };
                        _i = 0, executeOrderList_1 = executeOrderList;
                        _c.label = 1;
                    case 1:
                        if (!(_i < executeOrderList_1.length)) return [3 /*break*/, 5];
                        serverName = executeOrderList_1[_i];
                        logger_1.logger.info("Start executing project " + serverName);
                        return [4 /*yield*/, this.parse.clearGlobalKey('this').init(tempData)];
                    case 2:
                        parsedObj = _c.sent();
                        projectConfig = utils_1.getProjectConfig(parsedObj.realVariables, serverName, globalArgs);
                        return [4 /*yield*/, new component_1.default({
                                projectConfig: projectConfig,
                                method: method,
                                args: args,
                                spath: spath,
                                serverName: serverName,
                                globalArgs: globalArgs,
                                parse: this.parse,
                            }).init()];
                    case 3:
                        outputData = _c.sent();
                        tempData.services[serverName] = { output: outputData };
                        result[serverName] = outputData;
                        logger_1.logger.info("Project " + serverName + " successfully to execute \n\t");
                        _c.label = 4;
                    case 4:
                        _i++;
                        return [3 /*break*/, 1];
                    case 5:
                        lodash_1.keys(result).length === 0
                            ? logger_1.logger.log("End of method: " + method, 'green')
                            : logger_1.logger.output(result);
                        return [2 /*return*/, result];
                }
            });
        });
    };
    ExecCommand.prototype.warnEnvironmentVariables = function (realVariables) {
        return __awaiter(this, void 0, void 0, function () {
            var services, envObj, key, environmentVariables, keys, key;
            return __generator(this, function (_a) {
                services = realVariables === null || realVariables === void 0 ? void 0 : realVariables.services;
                if (lodash_1.isEmpty(services))
                    return [2 /*return*/];
                envObj = {};
                for (key in services) {
                    environmentVariables = lodash_1.get(services, [key, 'props', 'function', 'environmentVariables'], {});
                    envObj = Object.assign({}, envObj, environmentVariables);
                }
                keys = [];
                for (key in envObj) {
                    if (lodash_1.isNil(envObj[key])) {
                        keys.push(key);
                    }
                }
                keys.length > 0 &&
                    logger_1.logger.warn("The value of environment variable [" + keys.join(', ') + "] is undefined.");
                return [2 /*return*/];
            });
        });
    };
    return ExecCommand;
}());
function execCommand(configs) {
    return __awaiter(this, void 0, void 0, function () {
        return __generator(this, function (_a) {
            switch (_a.label) {
                case 0: return [4 /*yield*/, new ExecCommand(configs).init()];
                case 1: return [2 /*return*/, _a.sent()];
            }
        });
    });
}
exports.execCommand = execCommand;
