// @serverless-devs/core/dist/common/execCommand/utils.js
"use strict";
var __importDefault = (this && this.__importDefault) || function (mod) {
    return (mod && mod.__esModule) ? mod : { "default": mod };
};
Object.defineProperty(exports, "__esModule", { value: true });
exports.throwError = exports.getInputs = exports.getActions = exports.getCurrentPath = exports.getProjectConfig = exports.humanWarning = void 0;
var lodash_1 = require("lodash");
var path_1 = __importDefault(require("path"));
var libs_1 = require("../../libs");
var logger_1 = require("../../logger");
var chalk_1 = __importDefault(require("chalk"));
function humanWarning(tips) {
    logger_1.logger.log("\n" + chalk_1.default.hex('#000').bgYellow('WARNING:') + "\n======================\n" + libs_1.makeUnderLine(tips) + "\n", 'yellow');
}
exports.humanWarning = humanWarning;
function getProjectConfig(configs, serviceName, globalArgs) {
    var services = lodash_1.get(configs, 'services', {});
    var data = lodash_1.get(services, serviceName, {});
    var provider = data.provider || configs.provider;
    var access = data.access || configs.access;
    return lodash_1.assign({}, data, {
        access: (globalArgs === null || globalArgs === void 0 ? void 0 : globalArgs.access) || access,
        provider: provider,
        appName: configs.name,
        serviceName: serviceName,
    });
}
exports.getProjectConfig = getProjectConfig;
function getCurrentPath(p, spath) {
    if (p === void 0) { p = './'; }
    if (path_1.default.isAbsolute(p))
        return p;
    var dir = path_1.default.dirname(spath);
    return p ? path_1.default.join(dir, p) : dir;
}
exports.getCurrentPath = getCurrentPath;
function getActions(configs, _a) {
    var method = _a.method, spath = _a.spath;
    function validate(hook) {
        if ('run' in hook && !('component' in hook) && !('plugin' in hook))
            return 'run';
        if ('component' in hook && !('run' in hook) && !('plugin' in hook))
            return 'component';
        if ('plugin' in hook && !('run' in hook) && !('component' in hook))
            return 'plugin';
        throw new Error(JSON.stringify({
            message: 'actions configuration is incorrect.',
            tips: 'Please check the configuration of actions.',
        }));
    }
    var actions = configs.actions;
    if (lodash_1.isEmpty(actions))
        return;
    var hooks = [];
    var keyList = lodash_1.keys(actions);
    for (var _i = 0, keyList_1 = keyList; _i < keyList_1.length; _i++) {
        var actionKey = keyList_1[_i];
        var hookList = actions[actionKey];
        var _b = lodash_1.split(actionKey, '-'), start = _b[0], end = _b[1];
        if (end === method) {
            for (var _c = 0, hookList_1 = hookList; _c < hookList_1.length; _c++) {
                var hookDetail = hookList_1[_c];
                var type = validate(hookDetail);
                if (type === 'run') {
                    var obj = {
                        type: type,
                        value: hookDetail[type],
                        path: getCurrentPath(hookDetail.path, spath),
                        pre: start === 'pre' ? true : false,
                    };
                    hooks.push(obj);
                }
                if (type === 'component') {
                    var obj = {
                        type: type,
                        value: hookDetail[type],
                        pre: start === 'pre' ? true : false,
                    };
                    hooks.push(obj);
                }
                if (type === 'plugin') {
                    var obj = {
                        type: type,
                        value: hookDetail[type],
                        pre: start === 'pre' ? true : false,
                        args: hookDetail.args,
                    };
                    hooks.push(obj);
                }
            }
        }
    }
    return hooks;
}
exports.getActions = getActions;
function getInputs(configs, _a) {
    var method = _a.method, args = _a.args, spath = _a.spath, serverName = _a.serverName, output = _a.output;
    var argsObj = lodash_1.filter(args, function (o) { return !lodash_1.includes([serverName, method], o); });
    var inputs = {
        props: configs.props,
        credentials: configs.credentials,
        appName: configs.appName,
        project: {
            component: configs.component,
            access: configs.access,
            projectName: configs.serviceName,
            provider: configs.provider,
        },
        command: method,
        args: lodash_1.join(argsObj, ' '),
        argsObj: argsObj,
        path: {
            configPath: spath,
        },
        output: output,
    };
    return inputs;
}
exports.getInputs = getInputs;
function throwError(params) {
    var error = params.error, serviceName = params.serviceName;
    var jsonMsg;
    try {
        jsonMsg = JSON.parse(error.message);
    }
    catch (error) { }
    if (jsonMsg && jsonMsg.tips) {
        throw new Error(JSON.stringify({
            code: 101,
            message: jsonMsg.message,
            tips: jsonMsg.tips,
            prefix: "Project " + serviceName + " failed to execute:",
        }));
    }
    else {
        throw new Error(JSON.stringify({
            code: 101,
            message: error.message,
            stack: error.stack,
            prefix: "Project " + serviceName + " failed to execute:",
        }));
    }
}
exports.throwError = throwError;
