// @serverless-devs/core/dist/common/load/loadApplication.js
"use strict";
var __assign = (this && this.__assign) || function () {
    __assign = Object.assign || function(t) {
        for (var s, i = 1, n = arguments.length; i < n; i++) {
            s = arguments[i];
            for (var p in s) if (Object.prototype.hasOwnProperty.call(s, p))
                t[p] = s[p];
        }
        return t;
    };
    return __assign.apply(this, arguments);
};
var __createBinding = (this && this.__createBinding) || (Object.create ? (function(o, m, k, k2) {
    if (k2 === undefined) k2 = k;
    Object.defineProperty(o, k2, { enumerable: true, get: function() { return m[k]; } });
}) : (function(o, m, k, k2) {
    if (k2 === undefined) k2 = k;
    o[k2] = m[k];
}));
var __setModuleDefault = (this && this.__setModuleDefault) || (Object.create ? (function(o, v) {
    Object.defineProperty(o, "default", { enumerable: true, value: v });
}) : function(o, v) {
    o["default"] = v;
});
var __importStar = (this && this.__importStar) || function (mod) {
    if (mod && mod.__esModule) return mod;
    var result = {};
    if (mod != null) for (var k in mod) if (k !== "default" && Object.prototype.hasOwnProperty.call(mod, k)) __createBinding(result, mod, k);
    __setModuleDefault(result, mod);
    return result;
};
var __awaiter = (this && this.__awaiter) || function (thisArg, _arguments, P, generator) {
    function adopt(value) { return value instanceof P ? value : new P(function (resolve) { resolve(value); }); }
    return new (P || (P = Promise))(function (resolve, reject) {
        function fulfilled(value) { try { step(generator.next(value)); } catch (e) { reject(e); } }
        function rejected(value) { try { step(generator["throw"](value)); } catch (e) { reject(e); } }
        function step(result) { result.done ? resolve(result.value) : adopt(result.value).then(fulfilled, rejected); }
        step((generator = generator.apply(thisArg, _arguments || [])).next());
    });
};
var __generator = (this && this.__generator) || function (thisArg, body) {
    var _ = { label: 0, sent: function() { if (t[0] & 1) throw t[1]; return t[1]; }, trys: [], ops: [] }, f, y, t, g;
    return g = { next: verb(0), "throw": verb(1), "return": verb(2) }, typeof Symbol === "function" && (g[Symbol.iterator] = function() { return this; }), g;
    function verb(n) { return function (v) { return step([n, v]); }; }
    function step(op) {
        if (f) throw new TypeError("Generator is already executing.");
        while (_) try {
            if (f = 1, y && (t = op[0] & 2 ? y["return"] : op[0] ? y["throw"] || ((t = y["return"]) && t.call(y), 0) : y.next) && !(t = t.call(y, op[1])).done) return t;
            if (y = 0, t) op = [op[0] & 2, t.value];
            switch (op[0]) {
                case 0: case 1: t = op; break;
                case 4: _.label++; return { value: op[1], done: false };
                case 5: _.label++; y = op[1]; op = [0]; continue;
                case 7: op = _.ops.pop(); _.trys.pop(); continue;
                default:
                    if (!(t = _.trys, t = t.length > 0 && t[t.length - 1]) && (op[0] === 6 || op[0] === 2)) { _ = 0; continue; }
                    if (op[0] === 3 && (!t || (op[1] > t[0] && op[1] < t[3]))) { _.label = op[1]; break; }
                    if (op[0] === 6 && _.label < t[1]) { _.label = t[1]; t = op; break; }
                    if (t && _.label < t[2]) { _.label = t[2]; _.ops.push(op); break; }
                    if (t[2]) _.ops.pop();
                    _.trys.pop(); continue;
            }
            op = body.call(thisArg, _);
        } catch (e) { op = [6, e]; y = 0; } finally { f = t = 0; }
        if (op[0] & 5) throw op[1]; return { value: op[0] ? op[1] : void 0, done: true };
    }
};
var __importDefault = (this && this.__importDefault) || function (mod) {
    return (mod && mod.__esModule) ? mod : { "default": mod };
};
Object.defineProperty(exports, "__esModule", { value: true });
var service_1 = require("./service");
var constant_1 = require("../constant");
var path_1 = __importDefault(require("path"));
var downloadRequest_1 = __importDefault(require("../downloadRequest"));
var fs_extra_1 = __importDefault(require("fs-extra"));
var inquirer_1 = __importDefault(require("inquirer"));
var chalk_1 = __importDefault(require("chalk"));
var lodash_1 = __importStar(require("lodash"));
var rimraf_1 = __importDefault(require("rimraf"));
var installDependency_1 = __importDefault(require("../installDependency"));
var libs_1 = require("../../libs");
var credential_1 = require("../credential");
var utils_1 = require("./utils");
var parse_1 = __importDefault(require("./parse"));
var gray = chalk_1.default.hex('#8c8d91');
var artTemplate = require('art-template');
function tryfun(f) {
    return __awaiter(this, void 0, void 0, function () {
        var error_1;
        return __generator(this, function (_a) {
            switch (_a.label) {
                case 0:
                    _a.trys.push([0, 2, , 3]);
                    return [4 /*yield*/, f];
                case 1: return [2 /*return*/, _a.sent()];
                case 2:
                    error_1 = _a.sent();
                    return [3 /*break*/, 3];
                case 3: return [2 /*return*/];
            }
        });
    });
}
function preInit(_a) {
    var temporaryPath = _a.temporaryPath, applicationPath = _a.applicationPath;
    return __awaiter(this, void 0, void 0, function () {
        var baseChildComponent, tempObj, e_1;
        return __generator(this, function (_b) {
            switch (_b.label) {
                case 0:
                    _b.trys.push([0, 3, , 4]);
                    return [4 /*yield*/, require(path_1.default.join(temporaryPath, 'hook'))];
                case 1:
                    baseChildComponent = _b.sent();
                    tempObj = {
                        tempPath: temporaryPath,
                        targetPath: applicationPath,
                        downloadRequest: downloadRequest_1.default,
                        fse: fs_extra_1.default,
                        lodash: lodash_1.default,
                    };
                    return [4 /*yield*/, baseChildComponent.preInit(tempObj)];
                case 2:
                    _b.sent();
                    return [3 /*break*/, 4];
                case 3:
                    e_1 = _b.sent();
                    return [3 /*break*/, 4];
                case 4: return [2 /*return*/];
            }
        });
    });
}
function postInit(_a) {
    var temporaryPath = _a.temporaryPath, applicationPath = _a.applicationPath;
    return __awaiter(this, void 0, void 0, function () {
        var baseChildComponent, tempObj, e_2;
        return __generator(this, function (_b) {
            switch (_b.label) {
                case 0:
                    _b.trys.push([0, 3, , 4]);
                    return [4 /*yield*/, require(path_1.default.join(temporaryPath, 'hook'))];
                case 1:
                    baseChildComponent = _b.sent();
                    tempObj = {
                        tempPath: temporaryPath,
                        targetPath: applicationPath,
                        downloadRequest: downloadRequest_1.default,
                        fse: fs_extra_1.default,
                        lodash: lodash_1.default,
                    };
                    return [4 /*yield*/, baseChildComponent.postInit(tempObj)];
                case 2:
                    _b.sent();
                    return [3 /*break*/, 4];
                case 3:
                    e_2 = _b.sent();
                    return [3 /*break*/, 4];
                case 4: return [2 /*return*/];
            }
        });
    });
}
var LoadApplication = /** @class */ (function () {
    function LoadApplication(config) {
        this.config = config;
    }
    LoadApplication.prototype.byUrl = function () {
        return __awaiter(this, void 0, void 0, function () {
            var _a, source, registry, target, applicationPath;
            return __generator(this, function (_b) {
                switch (_b.label) {
                    case 0:
                        _a = this.config, source = _a.source, registry = _a.registry, target = _a.target;
                        applicationPath = path_1.default.resolve(target, source);
                        return [4 /*yield*/, downloadRequest_1.default(registry, applicationPath, {
                                extract: true,
                            })];
                    case 1:
                        _b.sent();
                        return [2 /*return*/, applicationPath];
                }
            });
        });
    };
    LoadApplication.prototype.loadType = function () {
        return __awaiter(this, void 0, void 0, function () {
            var registry;
            return __generator(this, function (_a) {
                switch (_a.label) {
                    case 0:
                        registry = this.config.registry;
                        if (!(registry === constant_1.RegistryEnum.serverless || registry === constant_1.RegistryEnum.serverlessOld)) return [3 /*break*/, 2];
                        return [4 /*yield*/, this.loadServerless()];
                    case 1: return [2 /*return*/, _a.sent()];
                    case 2:
                        if (!(registry === constant_1.RegistryEnum.github)) return [3 /*break*/, 4];
                        return [4 /*yield*/, this.loadGithub()];
                    case 3: return [2 /*return*/, _a.sent()];
                    case 4: return [2 /*return*/];
                }
            });
        });
    };
    LoadApplication.prototype.loadServerless = function () {
        return __awaiter(this, void 0, void 0, function () {
            var source, _a, provider, componentName, _b, name, version, zipball_url, result, findObj, result, newName, applicationPath;
            return __generator(this, function (_c) {
                switch (_c.label) {
                    case 0:
                        source = this.config.source.includes('/')
                            ? this.config.source
                            : "./" + this.config.source;
                        _a = source.split('/'), provider = _a[0], componentName = _a[1];
                        if (!componentName)
                            return [2 /*return*/];
                        _b = componentName.split('@'), name = _b[0], version = _b[1];
                        if (!version) return [3 /*break*/, 2];
                        return [4 /*yield*/, tryfun(service_1.getServerlessReleases(provider, name))];
                    case 1:
                        result = _c.sent();
                        if (!result)
                            return [2 /*return*/];
                        findObj = result.find(function (item) { return item.tag_name === version; });
                        if (!findObj)
                            return [2 /*return*/];
                        zipball_url = findObj.zipball_url;
                        return [3 /*break*/, 4];
                    case 2: return [4 /*yield*/, tryfun(service_1.getServerlessReleasesLatest(provider, name))];
                    case 3:
                        result = _c.sent();
                        if (!lodash_1.get(result, 'zipball_url'))
                            return [2 /*return*/];
                        zipball_url = result.zipball_url;
                        _c.label = 4;
                    case 4:
                        newName = this.config.name || name;
                        applicationPath = path_1.default.resolve(this.config.target, newName);
                        return [2 /*return*/, this.handleDecompressFile({
                                zipball_url: zipball_url,
                                applicationPath: applicationPath,
                                name: newName,
                            })];
                }
            });
        });
    };
    LoadApplication.prototype.loadGithub = function () {
        return __awaiter(this, void 0, void 0, function () {
            var _a, user, componentName, _b, name, version, zipball_url, result, findObj, result, newName, applicationPath;
            return __generator(this, function (_c) {
                switch (_c.label) {
                    case 0:
                        if (!this.config.source.includes('/'))
                            return [2 /*return*/];
                        _a = this.config.source.split('/'), user = _a[0], componentName = _a[1];
                        _b = componentName.split('@'), name = _b[0], version = _b[1];
                        if (!version) return [3 /*break*/, 2];
                        return [4 /*yield*/, tryfun(service_1.getGithubReleases(user, name))];
                    case 1:
                        result = _c.sent();
                        if (!result)
                            return [2 /*return*/];
                        findObj = result.find(function (item) { return item.tag_name === version; });
                        if (!findObj)
                            return [2 /*return*/];
                        zipball_url = findObj.zipball_url;
                        return [3 /*break*/, 4];
                    case 2: return [4 /*yield*/, tryfun(service_1.getGithubReleasesLatest(user, name))];
                    case 3:
                        result = _c.sent();
                        if (!lodash_1.get(result, 'zipball_url'))
                            return [2 /*return*/];
                        zipball_url = result.zipball_url;
                        _c.label = 4;
                    case 4:
                        newName = this.config.name || name;
                        applicationPath = path_1.default.join(this.config.target, newName);
                        return [2 /*return*/, this.handleDecompressFile({
                                zipball_url: zipball_url,
                                applicationPath: applicationPath,
                                name: newName,
                            })];
                }
            });
        });
    };
    LoadApplication.prototype.handleDecompressFile = function (_a) {
        var zipball_url = _a.zipball_url, applicationPath = _a.applicationPath, name = _a.name;
        return __awaiter(this, void 0, void 0, function () {
            var answer, temporaryPath, publishYamlData, _b;
            return __generator(this, function (_c) {
                switch (_c.label) {
                    case 0: return [4 /*yield*/, this.checkFileExists(applicationPath, name)];
                    case 1:
                        answer = _c.sent();
                        if (!answer)
                            return [2 /*return*/, applicationPath];
                        temporaryPath = "" + applicationPath + new Date().getTime();
                        return [4 /*yield*/, downloadRequest_1.default(zipball_url, temporaryPath, {
                                extract: true,
                                strip: 1,
                            })];
                    case 2:
                        _c.sent();
                        return [4 /*yield*/, preInit({ temporaryPath: temporaryPath, applicationPath: applicationPath })];
                    case 3:
                        _c.sent();
                        return [4 /*yield*/, libs_1.getYamlContent(path_1.default.join(temporaryPath, 'publish.yaml'))];
                    case 4:
                        publishYamlData = _c.sent();
                        if (!publishYamlData) return [3 /*break*/, 10];
                        fs_extra_1.default.copySync(temporaryPath + "/src", applicationPath);
                        rimraf_1.default.sync(temporaryPath);
                        if (!this.config.parameters) return [3 /*break*/, 6];
                        return [4 /*yield*/, this.initSconfigWithParam({ publishYamlData: publishYamlData, applicationPath: applicationPath })];
                    case 5:
                        _b = _c.sent();
                        return [3 /*break*/, 8];
                    case 6: return [4 /*yield*/, this.initSconfig({ publishYamlData: publishYamlData, applicationPath: applicationPath })];
                    case 7:
                        _b = _c.sent();
                        _c.label = 8;
                    case 8:
                        _b;
                        return [4 /*yield*/, this.initEnvConfig(applicationPath)];
                    case 9:
                        _c.sent();
                        return [3 /*break*/, 11];
                    case 10:
                        fs_extra_1.default.moveSync("" + temporaryPath, applicationPath);
                        _c.label = 11;
                    case 11: return [4 /*yield*/, this.needInstallDependency(applicationPath)];
                    case 12:
                        _c.sent();
                        return [4 /*yield*/, postInit({ temporaryPath: temporaryPath, applicationPath: applicationPath })];
                    case 13:
                        _c.sent();
                        return [2 /*return*/, applicationPath];
                }
            });
        });
    };
    LoadApplication.prototype.needInstallDependency = function (cwd) {
        return __awaiter(this, void 0, void 0, function () {
            var packageInfo, res;
            return __generator(this, function (_a) {
                switch (_a.label) {
                    case 0:
                        packageInfo = libs_1.readJsonFile(path_1.default.resolve(cwd, 'package.json'));
                        if (!packageInfo || !lodash_1.get(packageInfo, 'autoInstall', true))
                            return [2 /*return*/];
                        if (!process.env.skipPrompt) return [3 /*break*/, 2];
                        return [4 /*yield*/, tryfun(installDependency_1.default({ cwd: cwd, production: false }))];
                    case 1: return [2 /*return*/, _a.sent()];
                    case 2:
                        if (this.config.parameters)
                            return [2 /*return*/, true];
                        return [4 /*yield*/, inquirer_1.default.prompt([
                                {
                                    type: 'confirm',
                                    name: 'confirm',
                                    message: 'Do you want to install dependencies?',
                                    default: true,
                                },
                            ])];
                    case 3:
                        res = _a.sent();
                        if (!res.confirm) return [3 /*break*/, 5];
                        return [4 /*yield*/, tryfun(installDependency_1.default({ cwd: cwd, production: false }))];
                    case 4:
                        _a.sent();
                        _a.label = 5;
                    case 5: return [2 /*return*/];
                }
            });
        });
    };
    LoadApplication.prototype.initEnvConfig = function (appPath) {
        return __awaiter(this, void 0, void 0, function () {
            var envExampleFilePath, envConfig, templateKeys, promptOption, result, newEnvConfig;
            return __generator(this, function (_a) {
                switch (_a.label) {
                    case 0:
                        envExampleFilePath = path_1.default.resolve(appPath, '.env.example');
                        if (!fs_extra_1.default.existsSync(envExampleFilePath))
                            return [2 /*return*/];
                        envConfig = fs_extra_1.default.readFileSync(envExampleFilePath, 'utf-8');
                        templateKeys = utils_1.getTemplatekey(envConfig);
                        if (templateKeys.length === 0)
                            return [2 /*return*/];
                        promptOption = templateKeys.map(function (item) {
                            var name = item.name, desc = item.desc;
                            return {
                                type: 'input',
                                message: "please input " + (desc || name) + ":",
                                name: name,
                            };
                        });
                        return [4 /*yield*/, inquirer_1.default.prompt(promptOption)];
                    case 1:
                        result = _a.sent();
                        newEnvConfig = utils_1.replaceFun(envConfig, result);
                        fs_extra_1.default.unlink(envExampleFilePath);
                        fs_extra_1.default.writeFileSync(path_1.default.resolve(appPath, '.env'), newEnvConfig, 'utf-8');
                        return [2 /*return*/];
                }
            });
        });
    };
    LoadApplication.prototype.initSconfig = function (_a) {
        var publishYamlData = _a.publishYamlData, applicationPath = _a.applicationPath;
        return __awaiter(this, void 0, void 0, function () {
            var properties, requiredList, promptList, rangeList, key, ele, _loop_1, _i, rangeList_1, item, spath, credentialAliasList, _b, result, data, newData;
            return __generator(this, function (_c) {
                switch (_c.label) {
                    case 0:
                        properties = lodash_1.get(publishYamlData, 'Parameters.properties');
                        requiredList = lodash_1.get(publishYamlData, 'Parameters.required');
                        promptList = [];
                        if (properties) {
                            rangeList = [];
                            for (key in properties) {
                                ele = properties[key];
                                ele['_key'] = key;
                                rangeList.push(ele);
                            }
                            rangeList = lodash_1.sortBy(rangeList, function (o) { return o['x-range']; });
                            _loop_1 = function (item) {
                                var name_1 = item._key;
                                // 布尔类型
                                if (item.type === 'boolean') {
                                    promptList.push({
                                        type: 'confirm',
                                        name: name_1,
                                        message: item.description,
                                        default: item.default,
                                    });
                                }
                                else if (item.type === 'password') {
                                    // 密码类型
                                    promptList.push({
                                        type: 'password',
                                        name: name_1,
                                        message: item.description,
                                        default: item.default,
                                    });
                                }
                                else if (item.enum) {
                                    // 枚举类型
                                    promptList.push({
                                        type: 'list',
                                        name: name_1,
                                        prefix: item.description ? gray(item.description) + "\n" + chalk_1.default.green('?') : undefined,
                                        message: item.title,
                                        choices: item.enum,
                                        default: item.default,
                                    });
                                }
                                else if (item.type === 'string') {
                                    // 字符串类型
                                    promptList.push({
                                        type: 'input',
                                        message: item.title,
                                        name: name_1,
                                        prefix: item.description ? gray(item.description) + "\n" + chalk_1.default.green('?') : undefined,
                                        default: item.default,
                                        validate: function (input) {
                                            if (lodash_1.includes(requiredList, name_1)) {
                                                return input.length > 0 ? true : 'value cannot be empty.';
                                            }
                                            return true;
                                        },
                                    });
                                }
                            };
                            for (_i = 0, rangeList_1 = rangeList; _i < rangeList_1.length; _i++) {
                                item = rangeList_1[_i];
                                _loop_1(item);
                            }
                        }
                        spath = utils_1.getYamlPath(applicationPath, 's');
                        if (lodash_1.isEmpty(spath))
                            return [2 /*return*/];
                        _b = lodash_1.map;
                        return [4 /*yield*/, credential_1.getCredentialAliasList()];
                    case 1:
                        credentialAliasList = _b.apply(void 0, [_c.sent(), function (o) { return ({
                                name: o,
                                value: o,
                            }); }]);
                        result = {};
                        if (!lodash_1.isEmpty(credentialAliasList)) return [3 /*break*/, 5];
                        promptList.push({
                            type: 'confirm',
                            name: 'access',
                            message: 'create credential?',
                            default: true,
                        });
                        return [4 /*yield*/, inquirer_1.default.prompt(promptList)];
                    case 2:
                        result = _c.sent();
                        if (!(result === null || result === void 0 ? void 0 : result.access)) return [3 /*break*/, 4];
                        return [4 /*yield*/, credential_1.setCredential()];
                    case 3:
                        data = _c.sent();
                        result.access = data === null || data === void 0 ? void 0 : data.Alias;
                        _c.label = 4;
                    case 4: return [3 /*break*/, 7];
                    case 5:
                        promptList.push({
                            type: 'list',
                            name: 'access',
                            message: 'please select credential alias',
                            choices: lodash_1.concat(credentialAliasList, {
                                name: 'configure later.',
                                value: false,
                            }),
                        });
                        return [4 /*yield*/, inquirer_1.default.prompt(promptList)];
                    case 6:
                        result = _c.sent();
                        _c.label = 7;
                    case 7:
                        if ((result === null || result === void 0 ? void 0 : result.access) === false) {
                            result.access = '{{ access }}';
                        }
                        artTemplate.defaults.extname = path_1.default.extname(spath);
                        newData = artTemplate(spath, result);
                        // art 语法需要先解析在验证yaml内容
                        fs_extra_1.default.writeFileSync(spath, newData, 'utf-8');
                        // fix: Document with errors cannot be stringified
                        return [4 /*yield*/, libs_1.isYamlFile(spath)];
                    case 8:
                        // fix: Document with errors cannot be stringified
                        _c.sent();
                        newData = parse_1.default({ appName: this.config.appName }, newData);
                        fs_extra_1.default.writeFileSync(spath, newData, 'utf-8');
                        return [2 /*return*/];
                }
            });
        });
    };
    LoadApplication.prototype.initSconfigWithParam = function (_a) {
        var publishYamlData = _a.publishYamlData, applicationPath = _a.applicationPath;
        return __awaiter(this, void 0, void 0, function () {
            var spath, result, properties, requiredList, newObj, key, ele, accessObj, newData;
            return __generator(this, function (_b) {
                switch (_b.label) {
                    case 0:
                        spath = utils_1.getYamlPath(applicationPath, 's');
                        if (lodash_1.isEmpty(spath))
                            return [2 /*return*/];
                        result = this.config.parameters;
                        properties = lodash_1.get(publishYamlData, 'Parameters.properties');
                        requiredList = lodash_1.get(publishYamlData, 'Parameters.required', []);
                        newObj = {};
                        if (properties) {
                            for (key in properties) {
                                ele = properties[key];
                                if (result.hasOwnProperty(key)) {
                                    newObj[key] = result[key];
                                }
                                else if (ele.hasOwnProperty('default')) {
                                    newObj[key] = ele.default;
                                }
                                else if (lodash_1.includes(requiredList, key)) {
                                    throw new Error(key + " parameter is required.");
                                }
                            }
                        }
                        accessObj = this.config.access ? { access: this.config.access } : {};
                        artTemplate.defaults.extname = path_1.default.extname(spath);
                        newData = artTemplate(spath, __assign(__assign({}, newObj), accessObj));
                        fs_extra_1.default.writeFileSync(spath, newData, 'utf-8');
                        return [4 /*yield*/, libs_1.isYamlFile(spath)];
                    case 1:
                        _b.sent();
                        newData = parse_1.default({ appName: this.config.appName }, newData);
                        fs_extra_1.default.writeFileSync(spath, newData, 'utf-8');
                        return [2 /*return*/];
                }
            });
        });
    };
    LoadApplication.prototype.checkFileExists = function (filePath, fileName) {
        return __awaiter(this, void 0, void 0, function () {
            var res;
            return __generator(this, function (_a) {
                switch (_a.label) {
                    case 0:
                        if (process.env.skipPrompt)
                            return [2 /*return*/, true];
                        if (this.config.parameters)
                            return [2 /*return*/, true];
                        if (!fs_extra_1.default.existsSync(filePath)) return [3 /*break*/, 2];
                        return [4 /*yield*/, inquirer_1.default.prompt([
                                {
                                    type: 'confirm',
                                    name: 'confirm',
                                    message: "File " + fileName + " already exists, override this file ?",
                                    default: true,
                                },
                            ])];
                    case 1:
                        res = _a.sent();
                        return [2 /*return*/, res.confirm];
                    case 2: 
                    // 不存在文件，返回true表示需要覆盖
                    return [2 /*return*/, true];
                }
            });
        });
    };
    return LoadApplication;
}());
function loadApplication(oldsource, oldregistry, oldtarget) {
    return __awaiter(this, void 0, void 0, function () {
        var config, instance, appPath, registryFromSetConfig;
        return __generator(this, function (_a) {
            switch (_a.label) {
                case 0:
                    config = {};
                    if (typeof oldsource === 'string') {
                        config.source = oldsource;
                        config.registry = oldregistry;
                        config.target = oldtarget || libs_1.S_CURRENT;
                    }
                    else {
                        config.source = oldsource.source;
                        config.registry = oldsource.registry;
                        config.target = oldsource.target || libs_1.S_CURRENT;
                        config.name = oldsource.name;
                        config.parameters = oldsource.parameters;
                        config.appName = oldsource.appName;
                        config.access = oldsource.access;
                    }
                    instance = new LoadApplication(config);
                    if (!config.registry) return [3 /*break*/, 2];
                    if (!(config.registry !== constant_1.RegistryEnum.github && config.registry !== constant_1.RegistryEnum.serverless)) return [3 /*break*/, 2];
                    return [4 /*yield*/, instance.byUrl()];
                case 1: 
                // 支持 自定义
                return [2 /*return*/, _a.sent()];
                case 2:
                    if (!config.registry) return [3 /*break*/, 4];
                    return [4 /*yield*/, instance.loadType()];
                case 3:
                    appPath = _a.sent();
                    if (appPath)
                        return [2 /*return*/, appPath];
                    _a.label = 4;
                case 4: return [4 /*yield*/, libs_1.getSetConfig('registry')];
                case 5:
                    registryFromSetConfig = _a.sent();
                    if (!registryFromSetConfig) return [3 /*break*/, 7];
                    return [4 /*yield*/, instance.loadType()];
                case 6:
                    appPath = _a.sent();
                    if (appPath)
                        return [2 /*return*/, appPath];
                    _a.label = 7;
                case 7: return [4 /*yield*/, instance.loadServerless()];
                case 8:
                    appPath = _a.sent();
                    if (appPath)
                        return [2 /*return*/, appPath];
                    return [4 /*yield*/, instance.loadGithub()];
                case 9:
                    appPath = _a.sent();
                    if (appPath)
                        return [2 /*return*/, appPath];
                    if (!appPath) {
                        throw new Error(JSON.stringify({
                            message: "No " + config.source + " app found.",
                            tips: 'Please make sure the app name or source is correct, you can get applications on Serverless Regsitry: https://registry.serverless-devs.com',
                        }));
                    }
                    return [2 /*return*/];
            }
        });
    });
}
exports.default = loadApplication;
