// @serverless-devs/core/dist/common/load/loadComponent.js
"use strict";
var __awaiter = (this && this.__awaiter) || function (thisArg, _arguments, P, generator) {
    function adopt(value) { return value instanceof P ? value : new P(function (resolve) { resolve(value); }); }
    return new (P || (P = Promise))(function (resolve, reject) {
        function fulfilled(value) { try { step(generator.next(value)); } catch (e) { reject(e); } }
        function rejected(value) { try { step(generator["throw"](value)); } catch (e) { reject(e); } }
        function step(result) { result.done ? resolve(result.value) : adopt(result.value).then(fulfilled, rejected); }
        step((generator = generator.apply(thisArg, _arguments || [])).next());
    });
};
var __generator = (this && this.__generator) || function (thisArg, body) {
    var _ = { label: 0, sent: function() { if (t[0] & 1) throw t[1]; return t[1]; }, trys: [], ops: [] }, f, y, t, g;
    return g = { next: verb(0), "throw": verb(1), "return": verb(2) }, typeof Symbol === "function" && (g[Symbol.iterator] = function() { return this; }), g;
    function verb(n) { return function (v) { return step([n, v]); }; }
    function step(op) {
        if (f) throw new TypeError("Generator is already executing.");
        while (_) try {
            if (f = 1, y && (t = op[0] & 2 ? y["return"] : op[0] ? y["throw"] || ((t = y["return"]) && t.call(y), 0) : y.next) && !(t = t.call(y, op[1])).done) return t;
            if (y = 0, t) op = [op[0] & 2, t.value];
            switch (op[0]) {
                case 0: case 1: t = op; break;
                case 4: _.label++; return { value: op[1], done: false };
                case 5: _.label++; y = op[1]; op = [0]; continue;
                case 7: op = _.ops.pop(); _.trys.pop(); continue;
                default:
                    if (!(t = _.trys, t = t.length > 0 && t[t.length - 1]) && (op[0] === 6 || op[0] === 2)) { _ = 0; continue; }
                    if (op[0] === 3 && (!t || (op[1] > t[0] && op[1] < t[3]))) { _.label = op[1]; break; }
                    if (op[0] === 6 && _.label < t[1]) { _.label = t[1]; t = op; break; }
                    if (t && _.label < t[2]) { _.label = t[2]; _.ops.push(op); break; }
                    if (t[2]) _.ops.pop();
                    _.trys.pop(); continue;
            }
            op = body.call(thisArg, _);
        } catch (e) { op = [6, e]; y = 0; } finally { f = t = 0; }
        if (op[0] & 5) throw op[1]; return { value: op[0] ? op[1] : void 0, done: true };
    }
};
var __importDefault = (this && this.__importDefault) || function (mod) {
    return (mod && mod.__esModule) ? mod : { "default": mod };
};
Object.defineProperty(exports, "__esModule", { value: true });
exports.load = void 0;
var fs_extra_1 = __importDefault(require("fs-extra"));
var path_1 = __importDefault(require("path"));
var libs_1 = require("../../libs");
var service_1 = require("./service");
var constant_1 = require("../constant");
var utils_1 = require("./utils");
var downloadRequest_1 = __importDefault(require("../downloadRequest"));
var installDependency_1 = __importDefault(require("../installDependency"));
var lodash_1 = require("lodash");
var loadDevsCore_1 = require("./loadDevsCore");
var execDaemon_1 = require("../../execDaemon");
function tryfun(f) {
    return __awaiter(this, void 0, void 0, function () {
        var error_1;
        return __generator(this, function (_a) {
            switch (_a.label) {
                case 0:
                    _a.trys.push([0, 2, , 3]);
                    return [4 /*yield*/, f];
                case 1: return [2 /*return*/, _a.sent()];
                case 2:
                    error_1 = _a.sent();
                    return [3 /*break*/, 3];
                case 3: return [2 /*return*/];
            }
        });
    });
}
function preInit(_a) {
    var componentPath = _a.componentPath;
    return __awaiter(this, void 0, void 0, function () {
        var baseChildComponent, e_1;
        return __generator(this, function (_b) {
            switch (_b.label) {
                case 0:
                    _b.trys.push([0, 3, , 4]);
                    return [4 /*yield*/, require(path_1.default.join(componentPath, 'hook'))];
                case 1:
                    baseChildComponent = _b.sent();
                    return [4 /*yield*/, baseChildComponent.preInit({ componentPath: componentPath })];
                case 2:
                    _b.sent();
                    return [3 /*break*/, 4];
                case 3:
                    e_1 = _b.sent();
                    return [3 /*break*/, 4];
                case 4: return [2 /*return*/];
            }
        });
    });
}
function postInit(_a) {
    var componentPath = _a.componentPath;
    return __awaiter(this, void 0, void 0, function () {
        var baseChildComponent, e_2;
        return __generator(this, function (_b) {
            switch (_b.label) {
                case 0:
                    _b.trys.push([0, 3, , 4]);
                    return [4 /*yield*/, require(path_1.default.join(componentPath, 'hook'))];
                case 1:
                    baseChildComponent = _b.sent();
                    return [4 /*yield*/, baseChildComponent.postInit({ componentPath: componentPath })];
                case 2:
                    _b.sent();
                    return [3 /*break*/, 4];
                case 3:
                    e_2 = _b.sent();
                    return [3 /*break*/, 4];
                case 4: return [2 /*return*/];
            }
        });
    });
}
function getProviderComponent(source) {
    if (source.includes('/')) {
        return source.split('/');
    }
    var isFcComponent = lodash_1.find(constant_1.FC_COMPONENT, function (o) { return o === source; });
    return isFcComponent ? ['devsapp', source] : ['.', source];
}
function loadServerless(source, params) {
    return __awaiter(this, void 0, void 0, function () {
        var _a, provider, componentName, _b, name, version, componentPath, formatComponentName, filename, filename;
        return __generator(this, function (_c) {
            switch (_c.label) {
                case 0:
                    _a = getProviderComponent(source), provider = _a[0], componentName = _a[1];
                    if (!componentName)
                        return [2 /*return*/];
                    return [4 /*yield*/, utils_1.getComponentVersion(provider, componentName)];
                case 1:
                    _b = _c.sent(), name = _b[0], version = _b[1];
                    if (!version) return [3 /*break*/, 3];
                    formatComponentName = name + "@" + version;
                    filename = provider === '.' ? formatComponentName + ".zip" : provider + "_" + formatComponentName + ".zip";
                    return [4 /*yield*/, loadServerlessWithVersion({
                            provider: provider,
                            name: name,
                            componentName: formatComponentName,
                            version: version,
                            filename: filename,
                        })];
                case 2:
                    componentPath = _c.sent();
                    return [3 /*break*/, 5];
                case 3:
                    filename = provider === '.' ? name + ".zip" : provider + "_" + name + ".zip";
                    return [4 /*yield*/, loadServerlessWithNoVersion({ provider: provider, name: name, componentName: componentName, filename: filename })];
                case 4:
                    componentPath = _c.sent();
                    _c.label = 5;
                case 5:
                    if (!componentPath)
                        return [2 /*return*/];
                    return [4 /*yield*/, loadDevsCore_1.downLoadDesCore(componentPath)];
                case 6:
                    _c.sent();
                    return [4 /*yield*/, service_1.buildComponentInstance(componentPath, params)];
                case 7: return [2 /*return*/, _c.sent()];
            }
        });
    });
}
function loadServerlessWithVersion(_a) {
    var provider = _a.provider, name = _a.name, componentName = _a.componentName, version = _a.version, filename = _a.filename;
    return __awaiter(this, void 0, void 0, function () {
        var componentPath, lockPath, result, findObj, zipball_url;
        return __generator(this, function (_b) {
            switch (_b.label) {
                case 0:
                    componentPath = path_1.default.join(libs_1.getSComponentPath(), 'devsapp.cn', provider, componentName);
                    lockPath = path_1.default.resolve(componentPath, '.s.lock');
                    if (fs_extra_1.default.existsSync(lockPath)) {
                        return [2 /*return*/, componentPath];
                    }
                    return [4 /*yield*/, tryfun(service_1.getServerlessReleases(provider, name))];
                case 1:
                    result = _b.sent();
                    if (!result)
                        return [2 /*return*/];
                    findObj = result.find(function (item) { return item.tag_name === version; });
                    if (!findObj)
                        return [2 /*return*/];
                    zipball_url = findObj.zipball_url;
                    return [4 /*yield*/, downloadComponent({ zipball_url: zipball_url, filename: filename, componentPath: componentPath, lockPath: lockPath, version: version })];
                case 2:
                    _b.sent();
                    return [2 /*return*/, componentPath];
            }
        });
    });
}
function loadServerlessWithNoVersion(_a) {
    var provider = _a.provider, name = _a.name, componentName = _a.componentName, filename = _a.filename;
    return __awaiter(this, void 0, void 0, function () {
        var componentPath, lockPath, result, zipball_url, tag_name;
        return __generator(this, function (_b) {
            switch (_b.label) {
                case 0:
                    componentPath = path_1.default.join(libs_1.getSComponentPath(), 'devsapp.cn', provider, componentName);
                    lockPath = path_1.default.resolve(componentPath, '.s.lock');
                    if (fs_extra_1.default.existsSync(lockPath)) {
                        execDaemon_1.execDaemonWithTTL('loadComponent.js', {
                            componentPath: componentPath,
                            provider: provider,
                            name: name,
                            lockPath: lockPath,
                            registry: constant_1.RegistryEnum.serverless,
                        });
                        return [2 /*return*/, componentPath];
                    }
                    return [4 /*yield*/, tryfun(service_1.getServerlessReleasesLatest(provider, name))];
                case 1:
                    result = _b.sent();
                    if (!lodash_1.get(result, 'zipball_url'))
                        return [2 /*return*/];
                    zipball_url = result.zipball_url, tag_name = result.tag_name;
                    return [4 /*yield*/, downloadComponent({ zipball_url: zipball_url, filename: filename, componentPath: componentPath, lockPath: lockPath, version: tag_name })];
                case 2:
                    _b.sent();
                    return [2 /*return*/, componentPath];
            }
        });
    });
}
function downloadComponent(_a) {
    var zipball_url = _a.zipball_url, filename = _a.filename, componentPath = _a.componentPath, lockPath = _a.lockPath, version = _a.version;
    return __awaiter(this, void 0, void 0, function () {
        return __generator(this, function (_b) {
            switch (_b.label) {
                case 0: return [4 /*yield*/, downloadRequest_1.default(zipball_url, componentPath, {
                        filename: filename,
                        extract: true,
                        strip: 1,
                    })];
                case 1:
                    _b.sent();
                    return [4 /*yield*/, preInit({ componentPath: componentPath })];
                case 2:
                    _b.sent();
                    return [4 /*yield*/, installDependency_1.default({ cwd: componentPath, production: true })];
                case 3:
                    _b.sent();
                    fs_extra_1.default.writeFileSync(lockPath, JSON.stringify({ version: version }, null, 2));
                    return [4 /*yield*/, postInit({ componentPath: componentPath })];
                case 4:
                    _b.sent();
                    return [2 /*return*/];
            }
        });
    });
}
function loadGithub(source, params) {
    return __awaiter(this, void 0, void 0, function () {
        var _a, provider, componentName, _b, name, version, componentPath;
        return __generator(this, function (_c) {
            switch (_c.label) {
                case 0:
                    if (!source.includes('/'))
                        return [2 /*return*/];
                    _a = source.split('/'), provider = _a[0], componentName = _a[1];
                    if (!componentName)
                        return [2 /*return*/];
                    return [4 /*yield*/, utils_1.getComponentVersion(provider, componentName)];
                case 1:
                    _b = _c.sent(), name = _b[0], version = _b[1];
                    if (!version) return [3 /*break*/, 3];
                    return [4 /*yield*/, loadGithubWithVersion({
                            provider: provider,
                            name: name,
                            componentName: name + "@" + version,
                            version: version,
                        })];
                case 2:
                    componentPath = _c.sent();
                    return [3 /*break*/, 5];
                case 3: return [4 /*yield*/, loadGithubWithNoVersion({ provider: provider, name: name, componentName: componentName })];
                case 4:
                    componentPath = _c.sent();
                    _c.label = 5;
                case 5:
                    if (!componentPath)
                        return [2 /*return*/];
                    return [4 /*yield*/, loadDevsCore_1.downLoadDesCore(componentPath)];
                case 6:
                    _c.sent();
                    return [4 /*yield*/, service_1.buildComponentInstance(componentPath, params)];
                case 7: return [2 /*return*/, _c.sent()];
            }
        });
    });
}
function loadGithubWithVersion(_a) {
    var provider = _a.provider, name = _a.name, componentName = _a.componentName, version = _a.version;
    return __awaiter(this, void 0, void 0, function () {
        var componentPath, lockPath, result, findObj, zipball_url, filename;
        return __generator(this, function (_b) {
            switch (_b.label) {
                case 0:
                    componentPath = path_1.default.join(libs_1.getSComponentPath(), 'github.com', provider, componentName);
                    lockPath = path_1.default.resolve(componentPath, '.s.lock');
                    if (fs_extra_1.default.existsSync(lockPath)) {
                        return [2 /*return*/, componentPath];
                    }
                    return [4 /*yield*/, tryfun(service_1.getGithubReleases(provider, name))];
                case 1:
                    result = _b.sent();
                    if (!result)
                        return [2 /*return*/];
                    findObj = result.find(function (item) { return item.tag_name === version; });
                    if (!findObj)
                        return [2 /*return*/];
                    zipball_url = findObj.zipball_url;
                    filename = provider + "_" + componentName + ".zip";
                    return [4 /*yield*/, downloadComponent({ zipball_url: zipball_url, filename: filename, componentPath: componentPath, lockPath: lockPath, version: version })];
                case 2:
                    _b.sent();
                    return [2 /*return*/, componentPath];
            }
        });
    });
}
function loadGithubWithNoVersion(_a) {
    var provider = _a.provider, name = _a.name, componentName = _a.componentName;
    return __awaiter(this, void 0, void 0, function () {
        var componentPath, lockPath, result, zipball_url, tag_name, filename;
        return __generator(this, function (_b) {
            switch (_b.label) {
                case 0:
                    componentPath = path_1.default.join(libs_1.getSComponentPath(), 'github.com', provider, componentName);
                    lockPath = path_1.default.resolve(componentPath, '.s.lock');
                    if (fs_extra_1.default.existsSync(lockPath)) {
                        execDaemon_1.execDaemonWithTTL('loadComponent.js', {
                            componentPath: componentPath,
                            provider: provider,
                            name: name,
                            lockPath: lockPath,
                            registry: constant_1.RegistryEnum.github,
                        });
                        return [2 /*return*/, componentPath];
                    }
                    return [4 /*yield*/, tryfun(service_1.getGithubReleasesLatest(provider, name))];
                case 1:
                    result = _b.sent();
                    if (!lodash_1.get(result, 'zipball_url'))
                        return [2 /*return*/];
                    zipball_url = result.zipball_url, tag_name = result.tag_name;
                    // dev的tag有问题的，github下载地址重置
                    if (zipball_url.lastIndexOf('/dev') > 0) {
                        zipball_url = zipball_url
                            .replace(/^https:\/\/api.github.com/, 'https://github.com')
                            .replace('repos/', '')
                            .replace(/\/dev$/, '/refs/tags/dev');
                    }
                    filename = provider + "_" + componentName + "@" + tag_name + ".zip";
                    return [4 /*yield*/, downloadComponent({ zipball_url: zipball_url, filename: filename, componentPath: componentPath, lockPath: lockPath, version: tag_name })];
                case 2:
                    _b.sent();
                    return [2 /*return*/, componentPath];
            }
        });
    });
}
function loadType(source, registry, params) {
    return __awaiter(this, void 0, void 0, function () {
        return __generator(this, function (_a) {
            switch (_a.label) {
                case 0:
                    if (!(registry === constant_1.RegistryEnum.serverless || registry === constant_1.RegistryEnum.serverlessOld)) return [3 /*break*/, 2];
                    return [4 /*yield*/, loadServerless(source, params)];
                case 1: return [2 /*return*/, _a.sent()];
                case 2:
                    if (!(registry === constant_1.RegistryEnum.github)) return [3 /*break*/, 4];
                    return [4 /*yield*/, loadGithub(source, params)];
                case 3: return [2 /*return*/, _a.sent()];
                case 4: return [2 /*return*/];
            }
        });
    });
}
// TODO: 如何判断是一个组件
function isComponent(result) {
    return !!result;
}
function loadRemoteComponent(source, registry, params) {
    return __awaiter(this, void 0, void 0, function () {
        var result, registryFromSetConfig;
        return __generator(this, function (_a) {
            switch (_a.label) {
                case 0:
                    if (!registry) return [3 /*break*/, 2];
                    return [4 /*yield*/, loadType(source, registry, params)];
                case 1:
                    result = _a.sent();
                    if (isComponent(result))
                        return [2 /*return*/, result];
                    _a.label = 2;
                case 2: return [4 /*yield*/, libs_1.getSetConfig('registry')];
                case 3:
                    registryFromSetConfig = _a.sent();
                    if (!registryFromSetConfig) return [3 /*break*/, 5];
                    return [4 /*yield*/, loadType(source, registryFromSetConfig, params)];
                case 4:
                    result = _a.sent();
                    if (isComponent(result))
                        return [2 /*return*/, result];
                    _a.label = 5;
                case 5:
                    if (!(registryFromSetConfig !== constant_1.RegistryEnum.serverless &&
                        registryFromSetConfig !== constant_1.RegistryEnum.serverlessOld)) return [3 /*break*/, 7];
                    return [4 /*yield*/, loadServerless(source, params)];
                case 6:
                    result = _a.sent();
                    if (isComponent(result))
                        return [2 /*return*/, result];
                    _a.label = 7;
                case 7:
                    if (!(registryFromSetConfig !== constant_1.RegistryEnum.github)) return [3 /*break*/, 9];
                    return [4 /*yield*/, loadGithub(source, params)];
                case 8:
                    result = _a.sent();
                    if (isComponent(result))
                        return [2 /*return*/, result];
                    _a.label = 9;
                case 9:
                    if (!result) {
                        throw new Error("The " + source + " component was not found. Please make sure the component name or source is correct");
                    }
                    return [2 /*return*/];
            }
        });
    });
}
function loadComponent(source, registry, params) {
    return __awaiter(this, void 0, void 0, function () {
        return __generator(this, function (_a) {
            switch (_a.label) {
                case 0:
                    if (lodash_1.isEmpty(source)) {
                        throw new Error('Please check the value of source in loadComponent function.');
                    }
                    if (!fs_extra_1.default.existsSync(source)) return [3 /*break*/, 2];
                    return [4 /*yield*/, service_1.buildComponentInstance(source, params)];
                case 1: return [2 /*return*/, _a.sent()];
                case 2:
                    // js里引用下, 判断 registry 值是否 合法
                    if (registry) {
                        if (registry !== constant_1.RegistryEnum.github &&
                            registry !== constant_1.RegistryEnum.serverless &&
                            registry !== constant_1.RegistryEnum.serverlessOld) {
                            throw new Error("Please check the value of registry and set it to [" + constant_1.RegistryEnum.github + ", " + constant_1.RegistryEnum.serverless + "]");
                        }
                    }
                    return [4 /*yield*/, loadRemoteComponent(source, registry, params)];
                case 3: return [2 /*return*/, _a.sent()];
            }
        });
    });
}
exports.load = loadComponent;
exports.default = loadComponent;
