// @serverless-devs/core/dist/common/load/service.js
"use strict";
var __awaiter = (this && this.__awaiter) || function (thisArg, _arguments, P, generator) {
    function adopt(value) { return value instanceof P ? value : new P(function (resolve) { resolve(value); }); }
    return new (P || (P = Promise))(function (resolve, reject) {
        function fulfilled(value) { try { step(generator.next(value)); } catch (e) { reject(e); } }
        function rejected(value) { try { step(generator["throw"](value)); } catch (e) { reject(e); } }
        function step(result) { result.done ? resolve(result.value) : adopt(result.value).then(fulfilled, rejected); }
        step((generator = generator.apply(thisArg, _arguments || [])).next());
    });
};
var __generator = (this && this.__generator) || function (thisArg, body) {
    var _ = { label: 0, sent: function() { if (t[0] & 1) throw t[1]; return t[1]; }, trys: [], ops: [] }, f, y, t, g;
    return g = { next: verb(0), "throw": verb(1), "return": verb(2) }, typeof Symbol === "function" && (g[Symbol.iterator] = function() { return this; }), g;
    function verb(n) { return function (v) { return step([n, v]); }; }
    function step(op) {
        if (f) throw new TypeError("Generator is already executing.");
        while (_) try {
            if (f = 1, y && (t = op[0] & 2 ? y["return"] : op[0] ? y["throw"] || ((t = y["return"]) && t.call(y), 0) : y.next) && !(t = t.call(y, op[1])).done) return t;
            if (y = 0, t) op = [op[0] & 2, t.value];
            switch (op[0]) {
                case 0: case 1: t = op; break;
                case 4: _.label++; return { value: op[1], done: false };
                case 5: _.label++; y = op[1]; op = [0]; continue;
                case 7: op = _.ops.pop(); _.trys.pop(); continue;
                default:
                    if (!(t = _.trys, t = t.length > 0 && t[t.length - 1]) && (op[0] === 6 || op[0] === 2)) { _ = 0; continue; }
                    if (op[0] === 3 && (!t || (op[1] > t[0] && op[1] < t[3]))) { _.label = op[1]; break; }
                    if (op[0] === 6 && _.label < t[1]) { _.label = t[1]; t = op; break; }
                    if (t && _.label < t[2]) { _.label = t[2]; _.ops.push(op); break; }
                    if (t[2]) _.ops.pop();
                    _.trys.pop(); continue;
            }
            op = body.call(thisArg, _);
        } catch (e) { op = [6, e]; y = 0; } finally { f = t = 0; }
        if (op[0] & 5) throw op[1]; return { value: op[0] ? op[1] : void 0, done: true };
    }
};
var __importDefault = (this && this.__importDefault) || function (mod) {
    return (mod && mod.__esModule) ? mod : { "default": mod };
};
Object.defineProperty(exports, "__esModule", { value: true });
exports.getServerlessReleasesLatest = exports.getServerlessReleases = exports.getGithubReleasesLatest = exports.getGithubReleases = exports.buildComponentInstance = void 0;
var request_1 = __importDefault(require("../request"));
var utils_1 = require("../../libs/utils");
var path_1 = __importDefault(require("path"));
var fs_extra_1 = __importDefault(require("fs-extra"));
var lodash_1 = require("lodash");
var constant_1 = require("../constant");
var buildComponentInstance = function (componentPath, params) { return __awaiter(void 0, void 0, void 0, function () {
    var index, fsStat, packageInfo, tsconfigPath, tsconfigInfo, srcIndexPath, indexPath, requirePath, baseChildComponent, ChildComponent, componentInstance;
    return __generator(this, function (_a) {
        switch (_a.label) {
            case 0: return [4 /*yield*/, fs_extra_1.default.stat(componentPath)];
            case 1:
                fsStat = _a.sent();
                if (fsStat.isDirectory()) {
                    packageInfo = utils_1.readJsonFile(path_1.default.resolve(componentPath, 'package.json'));
                    // 首先寻找 package.json 文件下的 main
                    index = lodash_1.get(packageInfo, 'main');
                    // 其次检查 tsconfig.json 文件下的 outDir
                    if (!index) {
                        tsconfigPath = path_1.default.resolve(componentPath, 'tsconfig.json');
                        tsconfigInfo = utils_1.readJsonFile(tsconfigPath);
                        index = lodash_1.get(tsconfigInfo, 'compilerOptions.outDir');
                    }
                    // 其次寻找 src/index.js
                    if (!index) {
                        srcIndexPath = path_1.default.resolve(componentPath, './src/index.js');
                        if (fs_extra_1.default.existsSync(srcIndexPath)) {
                            index = './src/index.js';
                        }
                    }
                    // 最后寻找 ./index.js
                    if (!index) {
                        indexPath = path_1.default.resolve(componentPath, './index.js');
                        if (fs_extra_1.default.existsSync(indexPath)) {
                            index = './index.js';
                        }
                    }
                    if (!index) {
                        throw new Error('require不到组件, 请检查组件入口文件的设置是否设正确, 在当前目录下 首先寻找 package.json 文件下的 main, 其次寻找 tsconfig.json 文件下的 compilerOptions.outDir, 其次寻找 src/index.js, 最后寻找 index.js');
                    }
                }
                requirePath = fsStat.isFile() ? componentPath : path_1.default.resolve(componentPath, index);
                return [4 /*yield*/, require(requirePath)];
            case 2:
                baseChildComponent = _a.sent();
                ChildComponent = baseChildComponent.default
                    ? baseChildComponent.default
                    : baseChildComponent;
                try {
                    componentInstance = new ChildComponent(params);
                    if (componentInstance) {
                        componentInstance.__path = componentPath;
                    }
                    return [2 /*return*/, componentInstance];
                }
                catch (error) {
                    return [2 /*return*/, ChildComponent];
                }
                return [2 /*return*/];
        }
    });
}); };
exports.buildComponentInstance = buildComponentInstance;
var getGithubReleases = function (user, name) { return __awaiter(void 0, void 0, void 0, function () {
    return __generator(this, function (_a) {
        switch (_a.label) {
            case 0: return [4 /*yield*/, request_1.default(constant_1.RegistryEnum.github + "/" + user + "/" + name + "/releases", { ignoreError: true })];
            case 1: return [2 /*return*/, _a.sent()];
        }
    });
}); };
exports.getGithubReleases = getGithubReleases;
var getGithubReleasesLatest = function (user, name) { return __awaiter(void 0, void 0, void 0, function () {
    return __generator(this, function (_a) {
        switch (_a.label) {
            case 0: return [4 /*yield*/, request_1.default(constant_1.RegistryEnum.github + "/" + user + "/" + name + "/releases/latest", {
                    ignoreError: true,
                })];
            case 1: return [2 /*return*/, _a.sent()];
        }
    });
}); };
exports.getGithubReleasesLatest = getGithubReleasesLatest;
var getServerlessReleases = function (provider, name) { return __awaiter(void 0, void 0, void 0, function () {
    var url;
    return __generator(this, function (_a) {
        switch (_a.label) {
            case 0:
                url = provider === '.'
                    ? constant_1.RegistryEnum.serverless + "/" + name + "/releases"
                    : constant_1.RegistryEnum.serverless + "/" + provider + "/" + name + "/releases";
                return [4 /*yield*/, request_1.default(url, {
                        ignoreError: true,
                    })];
            case 1: return [2 /*return*/, _a.sent()];
        }
    });
}); };
exports.getServerlessReleases = getServerlessReleases;
var getServerlessReleasesLatest = function (provider, name) { return __awaiter(void 0, void 0, void 0, function () {
    var url;
    return __generator(this, function (_a) {
        switch (_a.label) {
            case 0:
                url = provider === '.'
                    ? constant_1.RegistryEnum.serverless + "/" + name + "/releases/latest"
                    : constant_1.RegistryEnum.serverless + "/" + provider + "/" + name + "/releases/latest";
                return [4 /*yield*/, request_1.default(url)];
            case 1: return [2 /*return*/, _a.sent()];
        }
    });
}); };
exports.getServerlessReleasesLatest = getServerlessReleasesLatest;
