// @serverless-devs/core/dist/common/load/utils.js
"use strict";
var __awaiter = (this && this.__awaiter) || function (thisArg, _arguments, P, generator) {
    function adopt(value) { return value instanceof P ? value : new P(function (resolve) { resolve(value); }); }
    return new (P || (P = Promise))(function (resolve, reject) {
        function fulfilled(value) { try { step(generator.next(value)); } catch (e) { reject(e); } }
        function rejected(value) { try { step(generator["throw"](value)); } catch (e) { reject(e); } }
        function step(result) { result.done ? resolve(result.value) : adopt(result.value).then(fulfilled, rejected); }
        step((generator = generator.apply(thisArg, _arguments || [])).next());
    });
};
var __generator = (this && this.__generator) || function (thisArg, body) {
    var _ = { label: 0, sent: function() { if (t[0] & 1) throw t[1]; return t[1]; }, trys: [], ops: [] }, f, y, t, g;
    return g = { next: verb(0), "throw": verb(1), "return": verb(2) }, typeof Symbol === "function" && (g[Symbol.iterator] = function() { return this; }), g;
    function verb(n) { return function (v) { return step([n, v]); }; }
    function step(op) {
        if (f) throw new TypeError("Generator is already executing.");
        while (_) try {
            if (f = 1, y && (t = op[0] & 2 ? y["return"] : op[0] ? y["throw"] || ((t = y["return"]) && t.call(y), 0) : y.next) && !(t = t.call(y, op[1])).done) return t;
            if (y = 0, t) op = [op[0] & 2, t.value];
            switch (op[0]) {
                case 0: case 1: t = op; break;
                case 4: _.label++; return { value: op[1], done: false };
                case 5: _.label++; y = op[1]; op = [0]; continue;
                case 7: op = _.ops.pop(); _.trys.pop(); continue;
                default:
                    if (!(t = _.trys, t = t.length > 0 && t[t.length - 1]) && (op[0] === 6 || op[0] === 2)) { _ = 0; continue; }
                    if (op[0] === 3 && (!t || (op[1] > t[0] && op[1] < t[3]))) { _.label = op[1]; break; }
                    if (op[0] === 6 && _.label < t[1]) { _.label = t[1]; t = op; break; }
                    if (t && _.label < t[2]) { _.label = t[2]; _.ops.push(op); break; }
                    if (t[2]) _.ops.pop();
                    _.trys.pop(); continue;
            }
            op = body.call(thisArg, _);
        } catch (e) { op = [6, e]; y = 0; } finally { f = t = 0; }
        if (op[0] & 5) throw op[1]; return { value: op[0] ? op[1] : void 0, done: true };
    }
};
var __importDefault = (this && this.__importDefault) || function (mod) {
    return (mod && mod.__esModule) ? mod : { "default": mod };
};
Object.defineProperty(exports, "__esModule", { value: true });
exports.getYamlPath = exports.getTemplatekey = exports.replaceFun = exports.getCoreVersion = exports.getComponentVersion = void 0;
var libs_1 = require("../../libs");
var lodash_1 = require("lodash");
var constant_1 = require("../../daemon/constant");
var path_1 = __importDefault(require("path"));
var fs_extra_1 = __importDefault(require("fs-extra"));
function getComponentVersion(provider, componentName) {
    return __awaiter(this, void 0, void 0, function () {
        var _a, name, version, core_load_serverless_devs_component, componentList, componentNames, obj, _i, componentList_1, item, _b, n, v, key_1;
        return __generator(this, function (_c) {
            _a = lodash_1.split(componentName, '@'), name = _a[0], version = _a[1];
            core_load_serverless_devs_component = process.env.core_load_serverless_devs_component;
            if (core_load_serverless_devs_component) {
                componentList = lodash_1.filter(lodash_1.split(core_load_serverless_devs_component, ';'), function (v) {
                    return lodash_1.includes(v, '@');
                });
                componentNames = [];
                obj = {};
                for (_i = 0, componentList_1 = componentList; _i < componentList_1.length; _i++) {
                    item = componentList_1[_i];
                    _b = lodash_1.split(item, '@'), n = _b[0], v = _b[1];
                    componentNames.push(n);
                    obj[n] = v;
                }
                key_1 = provider === '.' ? "devsapp/" + name : provider + "/" + name;
                if (lodash_1.find(componentNames, function (v) { return v === key_1; })) {
                    return [2 /*return*/, [name, obj[key_1]]];
                }
            }
            return [2 /*return*/, [name, version]];
        });
    });
}
exports.getComponentVersion = getComponentVersion;
function getCoreVersion() {
    return __awaiter(this, void 0, void 0, function () {
        return __generator(this, function (_a) {
            return [2 /*return*/, libs_1.isBetaS() ? 'dev' : constant_1.DEFAULT_CORE_VERSION];
        });
    });
}
exports.getCoreVersion = getCoreVersion;
function replaceFun(str, obj) {
    var reg = /\{\{(.*?)\}\}/g;
    var arr = str.match(reg);
    if (arr) {
        for (var i = 0; i < arr.length; i++) {
            var keyContent = arr[i].replace(/{{|}}/g, '');
            var realKey = lodash_1.trim(keyContent.split('|')[0]);
            str = str.replace(arr[i], obj[realKey]);
        }
    }
    return str;
}
exports.replaceFun = replaceFun;
function getTemplatekey(str) {
    var reg = /\{\{(.*?)\}\}/g;
    var arr = str.match(reg);
    if (!arr) {
        return [];
    }
    return arr
        .filter(function (result) { return result; })
        .map(function (matchValue) {
        var keyContent = matchValue.replace(/{{|}}/g, '');
        var realKey = keyContent.split('|');
        return {
            name: lodash_1.trim(realKey[0]),
            desc: lodash_1.trim(realKey[1]),
        };
    });
}
exports.getTemplatekey = getTemplatekey;
var getYamlPath = function (prePath, name) {
    var S_PATH1 = path_1.default.join(prePath, name + ".yaml");
    if (fs_extra_1.default.existsSync(S_PATH1))
        return S_PATH1;
    var S_PATH2 = path_1.default.join(prePath, name + ".yml");
    if (fs_extra_1.default.existsSync(S_PATH2))
        return S_PATH2;
};
exports.getYamlPath = getYamlPath;
