// @serverless-devs/core/dist/common/installDependency.js
"use strict";
var __awaiter = (this && this.__awaiter) || function (thisArg, _arguments, P, generator) {
    function adopt(value) { return value instanceof P ? value : new P(function (resolve) { resolve(value); }); }
    return new (P || (P = Promise))(function (resolve, reject) {
        function fulfilled(value) { try { step(generator.next(value)); } catch (e) { reject(e); } }
        function rejected(value) { try { step(generator["throw"](value)); } catch (e) { reject(e); } }
        function step(result) { result.done ? resolve(result.value) : adopt(result.value).then(fulfilled, rejected); }
        step((generator = generator.apply(thisArg, _arguments || [])).next());
    });
};
var __generator = (this && this.__generator) || function (thisArg, body) {
    var _ = { label: 0, sent: function() { if (t[0] & 1) throw t[1]; return t[1]; }, trys: [], ops: [] }, f, y, t, g;
    return g = { next: verb(0), "throw": verb(1), "return": verb(2) }, typeof Symbol === "function" && (g[Symbol.iterator] = function() { return this; }), g;
    function verb(n) { return function (v) { return step([n, v]); }; }
    function step(op) {
        if (f) throw new TypeError("Generator is already executing.");
        while (_) try {
            if (f = 1, y && (t = op[0] & 2 ? y["return"] : op[0] ? y["throw"] || ((t = y["return"]) && t.call(y), 0) : y.next) && !(t = t.call(y, op[1])).done) return t;
            if (y = 0, t) op = [op[0] & 2, t.value];
            switch (op[0]) {
                case 0: case 1: t = op; break;
                case 4: _.label++; return { value: op[1], done: false };
                case 5: _.label++; y = op[1]; op = [0]; continue;
                case 7: op = _.ops.pop(); _.trys.pop(); continue;
                default:
                    if (!(t = _.trys, t = t.length > 0 && t[t.length - 1]) && (op[0] === 6 || op[0] === 2)) { _ = 0; continue; }
                    if (op[0] === 3 && (!t || (op[1] > t[0] && op[1] < t[3]))) { _.label = op[1]; break; }
                    if (op[0] === 6 && _.label < t[1]) { _.label = t[1]; t = op; break; }
                    if (t && _.label < t[2]) { _.label = t[2]; _.ops.push(op); break; }
                    if (t[2]) _.ops.pop();
                    _.trys.pop(); continue;
            }
            op = body.call(thisArg, _);
        } catch (e) { op = [6, e]; y = 0; } finally { f = t = 0; }
        if (op[0] & 5) throw op[1]; return { value: op[0] ? op[1] : void 0, done: true };
    }
};
var __importDefault = (this && this.__importDefault) || function (mod) {
    return (mod && mod.__esModule) ? mod : { "default": mod };
};
Object.defineProperty(exports, "__esModule", { value: true });
var fs_extra_1 = __importDefault(require("fs-extra"));
var path_1 = __importDefault(require("path"));
var lodash_1 = require("lodash");
var execa_1 = __importDefault(require("execa"));
var report_1 = __importDefault(require("./report"));
var spinner_1 = __importDefault(require("./spinner"));
var state_1 = require("./state");
var crypto_js_1 = __importDefault(require("crypto-js"));
function checkYarn() {
    try {
        execa_1.default.sync('yarn', ['-v']);
        return true;
    }
    catch (error) {
        return false;
    }
}
var npmInstall = function (options) {
    if (options === void 0) { options = {}; }
    return __awaiter(void 0, void 0, void 0, function () {
        var showLoading, baseDir, npmList, production, spin, registry, client, errmsg;
        return __generator(this, function (_a) {
            showLoading = options.showLoading, baseDir = options.baseDir, npmList = options.npmList, production = options.production;
            if (showLoading) {
                spin = spinner_1.default('Dependencies installing...');
            }
            registry = options.registry ? " --registry=" + options.registry : '';
            try {
                client = checkYarn() ? 'yarn' : lodash_1.get(process.env, 'NPM_CLIENT', 'npm');
                execa_1.default.sync(client + " install " + (
                // eslint-disable-next-line no-nested-ternary
                npmList ? "" + npmList.join(' ') : production ? '--production' : '') + registry, { cwd: baseDir, shell: true, stdio: lodash_1.get(options, 'stdio', 'ignore') });
            }
            catch (error) {
                report_1.default({ type: 'installError', content: error.message + "||" + error.stack });
                errmsg = (error && error.message) || error;
                console.log(" - npm install err " + errmsg);
            }
            finally {
                if (showLoading) {
                    spin.stop();
                }
            }
            return [2 /*return*/];
        });
    });
};
function installDependency(options) {
    if (options === void 0) { options = {}; }
    return __awaiter(this, void 0, void 0, function () {
        var _a, cwd, _b, showLoading, _c, production, stdio, _d, graceInstall, packagePath, packageInfo, nonInstall, nodeModulePath;
        return __generator(this, function (_e) {
            switch (_e.label) {
                case 0:
                    _a = options.cwd, cwd = _a === void 0 ? process.cwd() : _a, _b = options.showLoading, showLoading = _b === void 0 ? true : _b, _c = options.production, production = _c === void 0 ? true : _c, stdio = options.stdio, _d = options.graceInstall, graceInstall = _d === void 0 ? false : _d;
                    packagePath = path_1.default.join(cwd, 'package.json');
                    if (!fs_extra_1.default.existsSync(packagePath))
                        return [2 /*return*/];
                    packageInfo = require(packagePath);
                    if (packageInfo.autoInstall === false)
                        return [2 /*return*/];
                    if (!graceInstall) return [3 /*break*/, 2];
                    return [4 /*yield*/, graceInstallDependency({ cwd: cwd, packageInfo: packageInfo, production: production })];
                case 1:
                    nonInstall = _e.sent();
                    if (nonInstall)
                        return [2 /*return*/];
                    return [3 /*break*/, 3];
                case 2:
                    nodeModulePath = path_1.default.join(cwd, 'node_modules');
                    if (fs_extra_1.default.existsSync(nodeModulePath))
                        return [2 /*return*/];
                    _e.label = 3;
                case 3: return [4 /*yield*/, npmInstall({
                        baseDir: cwd,
                        showLoading: showLoading,
                        production: production,
                        stdio: stdio,
                    })];
                case 4:
                    _e.sent();
                    return [2 /*return*/];
            }
        });
    });
}
function graceInstallDependency(_a) {
    var cwd = _a.cwd, packageInfo = _a.packageInfo, production = _a.production;
    return __awaiter(this, void 0, void 0, function () {
        var filename, sPath, encryptedContent, md5, data;
        return __generator(this, function (_b) {
            switch (_b.label) {
                case 0:
                    filename = "grace-install-dependency-for-" + lodash_1.get(packageInfo, 'name', 'package');
                    sPath = path_1.default.join(cwd, '.s');
                    encryptedContent = production ? lodash_1.get(packageInfo, 'dependencies', {}) : packageInfo;
                    md5 = crypto_js_1.default.MD5(JSON.stringify(encryptedContent)).toString();
                    return [4 /*yield*/, state_1.getState(filename, sPath)];
                case 1:
                    data = _b.sent();
                    if (lodash_1.get(data, 'md5') === md5) {
                        return [2 /*return*/, true];
                    }
                    return [4 /*yield*/, state_1.setState(filename, { md5: md5 }, sPath)];
                case 2:
                    _b.sent();
                    return [2 /*return*/, false];
            }
        });
    });
}
exports.default = installDependency;
