// @serverless-devs/core/dist/common/modifyYaml.js
"use strict";
var __assign = (this && this.__assign) || function () {
    __assign = Object.assign || function(t) {
        for (var s, i = 1, n = arguments.length; i < n; i++) {
            s = arguments[i];
            for (var p in s) if (Object.prototype.hasOwnProperty.call(s, p))
                t[p] = s[p];
        }
        return t;
    };
    return __assign.apply(this, arguments);
};
var __importDefault = (this && this.__importDefault) || function (mod) {
    return (mod && mod.__esModule) ? mod : { "default": mod };
};
Object.defineProperty(exports, "__esModule", { value: true });
var yaml_1 = __importDefault(require("yaml"));
var types_1 = require("yaml/types");
var lodash_1 = require("lodash");
var extend2_1 = __importDefault(require("extend2"));
var ModifyYaml = /** @class */ (function () {
    function ModifyYaml(json, yamlData) {
        this.globalKeys = [];
        this.doc = yaml_1.default.parseDocument(yamlData);
        // 新的json和原有的yaml数据进行合并，拿到一个最全的数据
        this.data = extend2_1.default(true, this.doc.toJSON(), json);
        if (lodash_1.get(json, 'vars')) {
            this.data = __assign(__assign({}, this.data), { vars: lodash_1.get(json, 'vars') });
        }
    }
    ModifyYaml.prototype.init = function () {
        var newDoc = yaml_1.default.parseDocument(yaml_1.default.stringify(this.data));
        var oldContents = this.doc.contents;
        this.setKey(newDoc, this.doc);
        // 对老的yaml收集注释信息等
        if (oldContents instanceof types_1.YAMLMap) {
            var items = oldContents.items;
            for (var _i = 0, items_1 = items; _i < items_1.length; _i++) {
                var item = items_1[_i];
                this.iteratorPair(item, item.key.value);
            }
        }
        // 对新的数据设置注释信息等
        var newContents = newDoc.contents;
        this.setKey(newContents, oldContents);
        if (newContents instanceof types_1.YAMLMap) {
            var items = newContents.items;
            for (var _a = 0, items_2 = items; _a < items_2.length; _a++) {
                var item = items_2[_a];
                this.setComment(item, item.key.value);
            }
        }
        return newDoc.toString();
    };
    ModifyYaml.prototype.addKey = function (item, key) {
        var findObj = lodash_1.find(this.globalKeys, function (o) { return o.key === key; });
        if (findObj)
            return;
        this.globalKeys.push({
            key: key,
            comment: item.comment,
            commentBefore: item.commentBefore,
            range: item.range,
            spaceBefore: item.spaceBefore,
        });
    };
    ModifyYaml.prototype.setKey = function (newVal, oldVal) {
        if (lodash_1.isEmpty(oldVal))
            return;
        newVal.comment = oldVal.comment;
        newVal.commentBefore = oldVal.commentBefore;
        newVal.range = oldVal.range;
        newVal.spaceBefore = oldVal.spaceBefore;
    };
    ModifyYaml.prototype.setComment = function (item, preKey) {
        var findObj = lodash_1.find(this.globalKeys, function (o) { return o.key === preKey; });
        this.setKey(item, findObj);
        if (item instanceof types_1.Pair) {
            if (item.value.type === 'MAP') {
                preKey += '.';
                for (var _i = 0, _a = item.value.items; _i < _a.length; _i++) {
                    var obj = _a[_i];
                    this.setComment(obj, preKey + obj.key.value);
                }
            }
            if (item.value.type === 'SEQ') {
                for (var index in item.value.items) {
                    var obj = item.value.items[index];
                    this.setComment(obj, preKey + ("[" + index + "]"));
                }
            }
            if (item.value.type === 'PLAIN') {
                this.setComment(item.value, preKey + '.value');
                return;
            }
        }
        if (item instanceof types_1.YAMLMap) {
            preKey += '.';
            for (var _b = 0, _c = item.items; _b < _c.length; _b++) {
                var obj = _c[_b];
                this.setComment(obj, preKey + obj.key.value);
            }
        }
    };
    ModifyYaml.prototype.iteratorPair = function (item, preKey) {
        this.addKey(item, preKey);
        if (item instanceof types_1.Pair) {
            if (item.value.type === 'MAP') {
                preKey += '.';
                for (var _i = 0, _a = item.value.items; _i < _a.length; _i++) {
                    var obj = _a[_i];
                    this.iteratorPair(obj, preKey + obj.key.value);
                }
                return;
            }
            if (item.value.type === 'SEQ') {
                for (var index in item.value.items) {
                    var obj = item.value.items[index];
                    this.iteratorPair(obj, preKey + ("[" + index + "]"));
                }
                return;
            }
            if (item.value.type === 'PLAIN') {
                this.iteratorPair(item.value, preKey + '.value');
                return;
            }
        }
        if (item instanceof types_1.YAMLMap) {
            preKey += '.';
            for (var _b = 0, _c = item.items; _b < _c.length; _b++) {
                var obj = _c[_b];
                this.iteratorPair(obj, preKey + obj.key.value);
            }
        }
    };
    return ModifyYaml;
}());
function modifyYaml(json, yamlData) {
    return new ModifyYaml(json, yamlData).init();
}
exports.default = modifyYaml;
