// @serverless-devs/core/dist/common/parseYaml.js
"use strict";
var __importDefault = (this && this.__importDefault) || function (mod) {
    return (mod && mod.__esModule) ? mod : { "default": mod };
};
Object.defineProperty(exports, "__esModule", { value: true });
var yaml_1 = __importDefault(require("yaml"));
var types_1 = require("yaml/types");
var lodash_1 = require("lodash");
var constant_1 = require("./constant");
var ParseYaml = /** @class */ (function () {
    function ParseYaml(data) {
        this.doc = yaml_1.default.parseDocument(data);
        this.yamlJson = this.doc.contents.toJSON();
    }
    ParseYaml.prototype.init = function () {
        var contents = this.doc.contents;
        if (contents instanceof types_1.YAMLMap) {
            var items = contents.items;
            for (var _i = 0, items_1 = items; _i < items_1.length; _i++) {
                var item = items_1[_i];
                this.iteratorPair(item);
            }
        }
        return this.doc.toJSON();
    };
    ParseYaml.prototype.iteratorPair = function (item) {
        if (item instanceof types_1.Pair) {
            var type = lodash_1.get(item, 'value.type');
            if (type === 'MAP' || type === 'SEQ') {
                for (var _i = 0, _a = item.value.items; _i < _a.length; _i++) {
                    var obj = _a[_i];
                    this.iteratorPair(obj);
                }
                return;
            }
            this.setPairValue(item);
        }
        if (item instanceof types_1.YAMLMap) {
            for (var _b = 0, _c = item.items; _b < _c.length; _b++) {
                var obj = _c[_b];
                this.iteratorPair(obj);
            }
        }
        if (item instanceof types_1.Scalar) {
            this.setScalarValue(item);
        }
    };
    ParseYaml.prototype.setPairValue = function (item) {
        var value = lodash_1.get(item, 'value.value');
        if (lodash_1.isBoolean(value) || lodash_1.isNumber(value) || lodash_1.isEmpty(value))
            return;
        var regResult = value.match(constant_1.COMMON_VARIABLE_TYPE_REG);
        if (regResult) {
            var tmp = this.getRealValue(value, regResult);
            item.value = yaml_1.default.createNode(tmp);
        }
    };
    ParseYaml.prototype.getRealValue = function (value, regResult) {
        var tmp = value;
        for (var _i = 0, regResult_1 = regResult; _i < regResult_1.length; _i++) {
            var iterator = regResult_1[_i];
            var realValue = lodash_1.get(this.yamlJson, iterator.replace(constant_1.COMMON_VARIABLE_TYPE_REG, '$1'), iterator);
            tmp = typeof realValue === 'string' ? tmp.replace(iterator, realValue) : realValue;
        }
        return tmp;
    };
    ParseYaml.prototype.setScalarValue = function (item) {
        if (lodash_1.isBoolean(item.value) || lodash_1.isNumber(item.value) || lodash_1.isEmpty(item.value))
            return;
        var regResult = item.value.match(constant_1.COMMON_VARIABLE_TYPE_REG);
        if (regResult) {
            item.value = this.getRealValue(item.value, regResult);
        }
    };
    return ParseYaml;
}());
function parseYaml(data) {
    return new ParseYaml(data).init();
}
exports.default = parseYaml;
