// @serverless-devs/core/dist/common/publishHelp.js
"use strict";
var __importDefault = (this && this.__importDefault) || function (mod) {
    return (mod && mod.__esModule) ? mod : { "default": mod };
};
Object.defineProperty(exports, "__esModule", { value: true });
var lodash_1 = __importDefault(require("lodash"));
var table_layout_1 = __importDefault(require("table-layout"));
var chalk_1 = require("chalk");
var utils_1 = require("../libs/utils");
var keyFn = function (list) { return lodash_1.default.first(lodash_1.default.keys(list)); };
var descFn = function (list) { return lodash_1.default.first(lodash_1.default.values(list)); };
var publishHelper = {
    /**
       *
       * Options: [
              {'--debug': 'Open debug model.'},
              {'--skip-actions': 'Skip the extends section.'},
              {'-t, --template <path>': 'Specify the template file.'},
              {'-a, --access <aliasName>': 'Specify the access alias name.'},
              {'-v, --version': 'Output the version number.'},
              {'-h, --help': 'Display help for command.'},
          ]
       * @param list
       */
    maxLen: function (list) {
        list = lodash_1.default.isArray(list) ? list : lodash_1.default.map(list, function (item, key) {
            var _a;
            return (_a = {}, _a[key] = item, _a);
        });
        return (lodash_1.default.map(list, function (item) { return keyFn(item); }).reduce(function (a, c) {
            return Math.max(a, c.length);
        }, 0) + 2);
    },
    /**
     *
     * @param list 初始化数据
     * @param title 名称
     * @param length 整体长度
     * @param leftPad 距离左侧长度
     * @returns
     */
    helpInfo: function (list, title, length, leftPad) {
        if (leftPad === void 0) { leftPad = 0; }
        if (lodash_1.default.isPlainObject(list) && lodash_1.default.isObject(descFn(list))) {
            return ("" + chalk_1.underline(chalk_1.bold(title)) +
                lodash_1.default.reduce(list, function (total, item, key) {
                    total += "\n" + publishHelper.helpInfo(item, key, length - 2, 2);
                    return total;
                }, ''));
        }
        list = lodash_1.default.isArray(list)
            ? lodash_1.default.map(list, function (item) { return ({
                command: [keyFn(item)],
                desc: descFn(item),
            }); })
            : lodash_1.default.map(list, function (item, key) { return ({ command: key, desc: item }); });
        if (lodash_1.default.isEmpty(list)) {
            return '';
        }
        var proxy = list.map(function (row) {
            return new Proxy(row, {
                get: function (target, property, receiver) {
                    if (property === 'desc') {
                        return "" + utils_1.makeUnderLine(target.desc);
                    }
                    else {
                        return Reflect.get(target, property, receiver);
                    }
                },
            });
        });
        return ("" + lodash_1.default.repeat(' ', leftPad) + (leftPad ? chalk_1.bold(title) : chalk_1.underline(chalk_1.bold(title))) + "\n" +
            new table_layout_1.default(proxy, {
                padding: { left: lodash_1.default.repeat(' ', leftPad + 2) },
                columns: [
                    {
                        name: 'command',
                        width: length + 2,
                    },
                ],
            }).toString());
        // `${_.repeat(' ', leftPad)}${leftPad?bold(title):underline(bold(title))}\n` + _.reduce(list, (total, item) => {
        //     let description = descFn(item);
        //     total+= (
        //         '  ' + _.padEnd(wrap(keyFn(item), {indent: _.repeat(' ', leftPad)}), length + 2) + wrap(description, {width: 80} ) + '\n');
        //     return total;
        //     },'')
    },
};
exports.default = publishHelper;
