// @serverless-devs/core/dist/common/request.js
"use strict";
var __assign = (this && this.__assign) || function () {
    __assign = Object.assign || function(t) {
        for (var s, i = 1, n = arguments.length; i < n; i++) {
            s = arguments[i];
            for (var p in s) if (Object.prototype.hasOwnProperty.call(s, p))
                t[p] = s[p];
        }
        return t;
    };
    return __assign.apply(this, arguments);
};
var __awaiter = (this && this.__awaiter) || function (thisArg, _arguments, P, generator) {
    function adopt(value) { return value instanceof P ? value : new P(function (resolve) { resolve(value); }); }
    return new (P || (P = Promise))(function (resolve, reject) {
        function fulfilled(value) { try { step(generator.next(value)); } catch (e) { reject(e); } }
        function rejected(value) { try { step(generator["throw"](value)); } catch (e) { reject(e); } }
        function step(result) { result.done ? resolve(result.value) : adopt(result.value).then(fulfilled, rejected); }
        step((generator = generator.apply(thisArg, _arguments || [])).next());
    });
};
var __generator = (this && this.__generator) || function (thisArg, body) {
    var _ = { label: 0, sent: function() { if (t[0] & 1) throw t[1]; return t[1]; }, trys: [], ops: [] }, f, y, t, g;
    return g = { next: verb(0), "throw": verb(1), "return": verb(2) }, typeof Symbol === "function" && (g[Symbol.iterator] = function() { return this; }), g;
    function verb(n) { return function (v) { return step([n, v]); }; }
    function step(op) {
        if (f) throw new TypeError("Generator is already executing.");
        while (_) try {
            if (f = 1, y && (t = op[0] & 2 ? y["return"] : op[0] ? y["throw"] || ((t = y["return"]) && t.call(y), 0) : y.next) && !(t = t.call(y, op[1])).done) return t;
            if (y = 0, t) op = [op[0] & 2, t.value];
            switch (op[0]) {
                case 0: case 1: t = op; break;
                case 4: _.label++; return { value: op[1], done: false };
                case 5: _.label++; y = op[1]; op = [0]; continue;
                case 7: op = _.ops.pop(); _.trys.pop(); continue;
                default:
                    if (!(t = _.trys, t = t.length > 0 && t[t.length - 1]) && (op[0] === 6 || op[0] === 2)) { _ = 0; continue; }
                    if (op[0] === 3 && (!t || (op[1] > t[0] && op[1] < t[3]))) { _.label = op[1]; break; }
                    if (op[0] === 6 && _.label < t[1]) { _.label = t[1]; t = op; break; }
                    if (t && _.label < t[2]) { _.label = t[2]; _.ops.push(op); break; }
                    if (t[2]) _.ops.pop();
                    _.trys.pop(); continue;
            }
            op = body.call(thisArg, _);
        } catch (e) { op = [6, e]; y = 0; } finally { f = t = 0; }
        if (op[0] & 5) throw op[1]; return { value: op[0] ? op[1] : void 0, done: true };
    }
};
var __rest = (this && this.__rest) || function (s, e) {
    var t = {};
    for (var p in s) if (Object.prototype.hasOwnProperty.call(s, p) && e.indexOf(p) < 0)
        t[p] = s[p];
    if (s != null && typeof Object.getOwnPropertySymbols === "function")
        for (var i = 0, p = Object.getOwnPropertySymbols(s); i < p.length; i++) {
            if (e.indexOf(p[i]) < 0 && Object.prototype.propertyIsEnumerable.call(s, p[i]))
                t[p[i]] = s[p[i]];
        }
    return t;
};
var __importDefault = (this && this.__importDefault) || function (mod) {
    return (mod && mod.__esModule) ? mod : { "default": mod };
};
Object.defineProperty(exports, "__esModule", { value: true });
exports.reportError = void 0;
var got_1 = __importDefault(require("@serverless-devs/got"));
var spinner_1 = __importDefault(require("./spinner"));
var logger_1 = require("../logger");
var report_1 = __importDefault(require("../common/report"));
function reportError(config) {
    report_1.default({
        type: 'networkError',
        content: config.requestUrl + "||" + config.statusCode + "||" + config.errorMsg,
    });
}
exports.reportError = reportError;
function request(url, options) {
    return __awaiter(this, void 0, void 0, function () {
        var _a, _b, method, params, bodyFromOptions, _c, hint, _d, json, _e, ignoreError, rest, loading, success, error, vm, result, errorMessage, isGet, e_1, statusCode, body;
        var _f;
        return __generator(this, function (_g) {
            switch (_g.label) {
                case 0:
                    _a = options || {}, _b = _a.method, method = _b === void 0 ? 'get' : _b, params = _a.params, bodyFromOptions = _a.body, _c = _a.hint, hint = _c === void 0 ? {} : _c, _d = _a.json, json = _d === void 0 ? true : _d, _e = _a.ignoreError, ignoreError = _e === void 0 ? false : _e, rest = __rest(_a, ["method", "params", "body", "hint", "json", "ignoreError"]);
                    loading = hint.loading, success = hint.success, error = hint.error;
                    vm = null;
                    result = null;
                    errorMessage = function (code, message) {
                        return "Url:" + url + "\n,params: " + JSON.stringify(options) + "\n,ErrorMessage:" + message + "\n, Code: " + code;
                    };
                    loading && (vm = spinner_1.default(loading));
                    _g.label = 1;
                case 1:
                    _g.trys.push([1, 3, , 4]);
                    isGet = method.toUpperCase() === 'GET';
                    logger_1.logger.debug("URL: " + url);
                    return [4 /*yield*/, got_1.default(url, __assign(__assign((_f = { method: method }, _f[isGet ? 'query' : 'body'] = isGet ? params : bodyFromOptions, _f.json = json, _f), rest), { rejectUnauthorized: false }))];
                case 2:
                    result = _g.sent();
                    loading && vm.stop();
                    return [3 /*break*/, 4];
                case 3:
                    e_1 = _g.sent();
                    loading && vm.stop();
                    if (!ignoreError) {
                        spinner_1.default(e_1.message).fail();
                        reportError({
                            requestUrl: url,
                            statusCode: e_1.code,
                            errorMsg: e_1.message,
                        });
                        throw new Error(errorMessage(e_1.statusCode, e_1.message));
                    }
                    return [3 /*break*/, 4];
                case 4:
                    statusCode = result.statusCode, body = result.body;
                    if (statusCode !== 200) {
                        error && spinner_1.default(error).fail();
                        if (!ignoreError) {
                            reportError({
                                requestUrl: url,
                                statusCode: statusCode,
                                errorMsg: 'System exception',
                            });
                            throw new Error(errorMessage(statusCode, 'System exception'));
                        }
                    }
                    else if (body.Error) {
                        error && spinner_1.default(error).fail();
                        if (!ignoreError) {
                            reportError({
                                requestUrl: url,
                                statusCode: body.Error.Code,
                                errorMsg: body.Error.Message,
                            });
                            throw new Error(errorMessage(body.Error.Code, body.Error.Message));
                        }
                    }
                    success && spinner_1.default(success).succeed();
                    return [2 /*return*/, body.Response || body];
            }
        });
    });
}
exports.default = request;
